{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pickle as pkl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/usrvol/mappings/classes.txt', 'r') as f:\n",
    "    classes = f.read().splitlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/usrvol/mappings/classes.pickle', 'wb') as f:\n",
    "    pkl.dump(classes, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/usrvol/mappings/mapping_definitivo.pickle', 'rb') as f:\n",
    "    mapping = pkl.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NONE'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping['T194']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping['T194'] = 'NONE'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/usrvol/mappings/mapping_definitivo.pickle', 'wb') as f:\n",
    "    pkl.dump(mapping, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Vocabulario y vocab_mapper obtenidos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle as pkl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/usrvol/mappings/categorical/vocab_mapper.pickle', 'rb') as f:\n",
    "    vocab_mapper = pkl.load(f)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 0,\n",
       " 1: 1,\n",
       " 2: 2,\n",
       " 3: 3,\n",
       " 4: 4,\n",
       " 5: 5,\n",
       " 6: 6,\n",
       " 7: 7,\n",
       " 8: 8,\n",
       " 9: 9,\n",
       " 10: 10,\n",
       " 11: 11,\n",
       " 12: 12,\n",
       " 13: 13,\n",
       " 14: 14,\n",
       " 15: 15,\n",
       " 16: 16,\n",
       " 17: 17,\n",
       " 18: 18,\n",
       " 19: 19,\n",
       " 20: 20,\n",
       " 21: 21,\n",
       " 22: 22,\n",
       " 23: 23,\n",
       " 24: 24,\n",
       " 25: 25,\n",
       " 26: 26,\n",
       " 27: 27,\n",
       " 28: 28,\n",
       " 29: 29,\n",
       " 30: 30,\n",
       " 31: 31,\n",
       " 32: 32,\n",
       " 33: 33,\n",
       " 34: 34,\n",
       " 35: 35,\n",
       " 36: 36,\n",
       " 37: 37,\n",
       " 38: 38,\n",
       " 39: 39,\n",
       " 40: 40,\n",
       " 41: 41,\n",
       " 42: 42,\n",
       " 43: 43,\n",
       " 44: 44,\n",
       " 45: 45,\n",
       " 46: 46,\n",
       " 47: 47,\n",
       " 48: 48,\n",
       " 49: 49,\n",
       " 50: 50,\n",
       " 51: 51,\n",
       " 52: 52,\n",
       " 53: 53,\n",
       " 54: 54,\n",
       " 55: 55,\n",
       " 56: 56,\n",
       " 57: 57,\n",
       " 58: 58,\n",
       " 59: 59,\n",
       " 60: 60,\n",
       " 61: 61,\n",
       " 62: 62,\n",
       " 63: 63,\n",
       " 64: 64,\n",
       " 65: 65,\n",
       " 66: 66,\n",
       " 67: 67,\n",
       " 68: 68,\n",
       " 69: 69,\n",
       " 70: 70,\n",
       " 71: 71,\n",
       " 72: 72,\n",
       " 73: 73,\n",
       " 74: 74,\n",
       " 75: 75,\n",
       " 76: 76,\n",
       " 77: 77,\n",
       " 78: 78,\n",
       " 79: 79,\n",
       " 80: 80,\n",
       " 81: 81,\n",
       " 82: 82,\n",
       " 83: 83,\n",
       " 84: 84,\n",
       " 85: 85,\n",
       " 86: 86,\n",
       " 87: 87,\n",
       " 88: 88,\n",
       " 89: 89,\n",
       " 90: 90,\n",
       " 91: 91,\n",
       " 92: 92,\n",
       " 93: 93,\n",
       " 94: 94,\n",
       " 95: 95,\n",
       " 96: 96,\n",
       " 97: 97,\n",
       " 98: 98,\n",
       " 99: 99,\n",
       " 100: 100,\n",
       " 101: 101,\n",
       " 102: 102,\n",
       " 103: 103,\n",
       " 104: 104,\n",
       " 105: 105,\n",
       " 106: 106,\n",
       " 107: 107,\n",
       " 108: 108,\n",
       " 109: 109,\n",
       " 110: 110,\n",
       " 111: 111,\n",
       " 112: 112,\n",
       " 113: 113,\n",
       " 114: 114,\n",
       " 115: 115,\n",
       " 116: 116,\n",
       " 117: 117,\n",
       " 118: 118,\n",
       " 119: 119,\n",
       " 120: 120,\n",
       " 121: 121,\n",
       " 122: 122,\n",
       " 123: 123,\n",
       " 124: 124,\n",
       " 125: 125,\n",
       " 126: 126,\n",
       " 127: 127,\n",
       " 128: 128,\n",
       " 129: 129,\n",
       " 130: 130,\n",
       " 131: 131,\n",
       " 132: 132,\n",
       " 133: 133,\n",
       " 134: 134,\n",
       " 135: 135,\n",
       " 136: 136,\n",
       " 137: 137,\n",
       " 138: 138,\n",
       " 139: 139,\n",
       " 140: 140,\n",
       " 141: 141,\n",
       " 142: 142,\n",
       " 143: 143,\n",
       " 144: 144,\n",
       " 145: 145,\n",
       " 146: 146,\n",
       " 147: 147,\n",
       " 148: 148,\n",
       " 149: 149,\n",
       " 150: 150,\n",
       " 151: 151,\n",
       " 152: 152,\n",
       " 153: 153,\n",
       " 154: 154,\n",
       " 155: 155,\n",
       " 156: 156,\n",
       " 157: 157,\n",
       " 158: 158,\n",
       " 159: 159,\n",
       " 160: 160,\n",
       " 161: 161,\n",
       " 162: 162,\n",
       " 163: 163,\n",
       " 164: 164,\n",
       " 165: 165,\n",
       " 166: 166,\n",
       " 167: 167,\n",
       " 168: 168,\n",
       " 169: 169,\n",
       " 170: 170,\n",
       " 171: 171,\n",
       " 172: 172,\n",
       " 173: 173,\n",
       " 174: 174,\n",
       " 175: 175,\n",
       " 176: 176,\n",
       " 177: 177,\n",
       " 178: 178,\n",
       " 179: 179,\n",
       " 180: 180,\n",
       " 181: 181,\n",
       " 182: 182,\n",
       " 183: 183,\n",
       " 184: 184,\n",
       " 185: 185,\n",
       " 186: 186,\n",
       " 187: 187,\n",
       " 188: 188,\n",
       " 189: 189,\n",
       " 190: 190,\n",
       " 191: 191,\n",
       " 192: 192,\n",
       " 193: 193,\n",
       " 194: 194,\n",
       " 195: 195,\n",
       " 196: 196,\n",
       " 197: 197,\n",
       " 198: 198,\n",
       " 199: 199,\n",
       " 200: 200,\n",
       " 201: 201,\n",
       " 202: 202,\n",
       " 203: 203,\n",
       " 204: 204,\n",
       " 205: 205,\n",
       " 206: 206,\n",
       " 207: 207,\n",
       " 208: 208,\n",
       " 209: 209,\n",
       " 210: 210,\n",
       " 211: 211,\n",
       " 212: 212,\n",
       " 213: 213,\n",
       " 214: 214,\n",
       " 215: 215,\n",
       " 216: 216,\n",
       " 217: 217,\n",
       " 218: 218,\n",
       " 219: 219,\n",
       " 220: 220,\n",
       " 221: 221,\n",
       " 222: 222,\n",
       " 223: 223,\n",
       " 224: 224,\n",
       " 225: 225,\n",
       " 226: 226,\n",
       " 227: 227,\n",
       " 228: 228,\n",
       " 229: 229,\n",
       " 230: 230,\n",
       " 231: 231,\n",
       " 232: 232,\n",
       " 233: 233,\n",
       " 234: 234,\n",
       " 235: 235,\n",
       " 236: 236,\n",
       " 237: 237,\n",
       " 238: 238,\n",
       " 239: 239,\n",
       " 240: 240,\n",
       " 241: 241,\n",
       " 242: 242,\n",
       " 243: 243,\n",
       " 244: 244,\n",
       " 245: 245,\n",
       " 246: 246,\n",
       " 247: 247,\n",
       " 248: 248,\n",
       " 249: 249,\n",
       " 250: 250,\n",
       " 251: 251,\n",
       " 252: 252,\n",
       " 253: 253,\n",
       " 254: 254,\n",
       " 255: 255,\n",
       " 256: 256,\n",
       " 257: 257,\n",
       " 258: 258,\n",
       " 259: 259,\n",
       " 260: 260,\n",
       " 261: 261,\n",
       " 262: 262,\n",
       " 263: 263,\n",
       " 264: 264,\n",
       " 265: 265,\n",
       " 266: 266,\n",
       " 267: 267,\n",
       " 268: 268,\n",
       " 269: 269,\n",
       " 270: 270,\n",
       " 271: 271,\n",
       " 272: 272,\n",
       " 273: 273,\n",
       " 274: 274,\n",
       " 275: 275,\n",
       " 276: 276,\n",
       " 277: 277,\n",
       " 278: 278,\n",
       " 279: 279,\n",
       " 280: 280,\n",
       " 281: 281,\n",
       " 282: 282,\n",
       " 283: 283,\n",
       " 284: 284,\n",
       " 285: 285,\n",
       " 286: 286,\n",
       " 287: 287,\n",
       " 288: 288,\n",
       " 289: 289,\n",
       " 290: 290,\n",
       " 291: 291,\n",
       " 292: 292,\n",
       " 293: 293,\n",
       " 294: 294,\n",
       " 295: 295,\n",
       " 296: 296,\n",
       " 297: 297,\n",
       " 298: 298,\n",
       " 299: 299,\n",
       " 300: 300,\n",
       " 301: 301,\n",
       " 302: 302,\n",
       " 303: 303,\n",
       " 304: 304,\n",
       " 305: 305,\n",
       " 306: 306,\n",
       " 307: 307,\n",
       " 308: 308,\n",
       " 309: 309,\n",
       " 310: 310,\n",
       " 311: 311,\n",
       " 312: 312,\n",
       " 313: 313,\n",
       " 314: 314,\n",
       " 315: 315,\n",
       " 316: 316,\n",
       " 317: 317,\n",
       " 318: 318,\n",
       " 319: 319,\n",
       " 320: 320,\n",
       " 321: 321,\n",
       " 322: 322,\n",
       " 323: 323,\n",
       " 324: 324,\n",
       " 325: 325,\n",
       " 326: 326,\n",
       " 327: 327,\n",
       " 328: 328,\n",
       " 329: 329,\n",
       " 330: 330,\n",
       " 331: 331,\n",
       " 332: 332,\n",
       " 333: 333,\n",
       " 334: 334,\n",
       " 335: 335,\n",
       " 336: 336,\n",
       " 337: 337,\n",
       " 338: 338,\n",
       " 339: 339,\n",
       " 340: 340,\n",
       " 341: 341,\n",
       " 342: 342,\n",
       " 343: 343,\n",
       " 344: 344,\n",
       " 345: 345,\n",
       " 346: 346,\n",
       " 347: 347,\n",
       " 348: 348,\n",
       " 349: 349,\n",
       " 350: 350,\n",
       " 351: 351,\n",
       " 352: 352,\n",
       " 353: 353,\n",
       " 354: 354,\n",
       " 355: 355,\n",
       " 356: 356,\n",
       " 357: 357,\n",
       " 358: 358,\n",
       " 359: 359,\n",
       " 360: 360,\n",
       " 361: 361,\n",
       " 362: 362,\n",
       " 363: 363,\n",
       " 364: 364,\n",
       " 365: 365,\n",
       " 366: 366,\n",
       " 367: 367,\n",
       " 368: 368,\n",
       " 369: 369,\n",
       " 370: 370,\n",
       " 371: 371,\n",
       " 372: 372,\n",
       " 373: 373,\n",
       " 374: 374,\n",
       " 375: 375,\n",
       " 376: 376,\n",
       " 377: 377,\n",
       " 378: 378,\n",
       " 379: 379,\n",
       " 380: 380,\n",
       " 381: 381,\n",
       " 382: 382,\n",
       " 383: 383,\n",
       " 384: 384,\n",
       " 385: 385,\n",
       " 386: 386,\n",
       " 387: 387,\n",
       " 388: 388,\n",
       " 389: 389,\n",
       " 390: 390,\n",
       " 391: 391,\n",
       " 392: 392,\n",
       " 393: 393,\n",
       " 394: 394,\n",
       " 395: 395,\n",
       " 396: 396,\n",
       " 397: 397,\n",
       " 398: 398,\n",
       " 399: 399,\n",
       " 400: 400,\n",
       " 401: 401,\n",
       " 402: 402,\n",
       " 403: 403,\n",
       " 404: 404,\n",
       " 405: 405,\n",
       " 406: 406,\n",
       " 407: 407,\n",
       " 408: 408,\n",
       " 409: 409,\n",
       " 410: 410,\n",
       " 411: 411,\n",
       " 412: 412,\n",
       " 413: 413,\n",
       " 414: 414,\n",
       " 415: 415,\n",
       " 416: 416,\n",
       " 417: 417,\n",
       " 418: 418,\n",
       " 419: 419,\n",
       " 420: 420,\n",
       " 421: 421,\n",
       " 422: 422,\n",
       " 423: 423,\n",
       " 424: 424,\n",
       " 425: 425,\n",
       " 426: 426,\n",
       " 427: 427,\n",
       " 428: 428,\n",
       " 429: 429,\n",
       " 430: 430,\n",
       " 431: 431,\n",
       " 432: 432,\n",
       " 433: 433,\n",
       " 434: 434,\n",
       " 435: 435,\n",
       " 436: 436,\n",
       " 437: 437,\n",
       " 438: 438,\n",
       " 439: 439,\n",
       " 440: 440,\n",
       " 441: 441,\n",
       " 442: 442,\n",
       " 443: 443,\n",
       " 444: 444,\n",
       " 445: 445,\n",
       " 446: 446,\n",
       " 447: 447,\n",
       " 448: 448,\n",
       " 449: 449,\n",
       " 450: 450,\n",
       " 451: 451,\n",
       " 452: 452,\n",
       " 453: 453,\n",
       " 454: 454,\n",
       " 455: 455,\n",
       " 456: 456,\n",
       " 457: 457,\n",
       " 458: 458,\n",
       " 459: 459,\n",
       " 460: 460,\n",
       " 461: 461,\n",
       " 462: 462,\n",
       " 463: 463,\n",
       " 464: 464,\n",
       " 465: 465,\n",
       " 466: 466,\n",
       " 467: 467,\n",
       " 468: 468,\n",
       " 469: 469,\n",
       " 470: 470,\n",
       " 471: 471,\n",
       " 472: 472,\n",
       " 473: 473,\n",
       " 474: 474,\n",
       " 475: 475,\n",
       " 476: 476,\n",
       " 477: 477,\n",
       " 478: 478,\n",
       " 479: 479,\n",
       " 480: 480,\n",
       " 481: 481,\n",
       " 482: 482,\n",
       " 483: 483,\n",
       " 484: 484,\n",
       " 485: 485,\n",
       " 486: 486,\n",
       " 487: 487,\n",
       " 488: 488,\n",
       " 489: 489,\n",
       " 490: 490,\n",
       " 491: 491,\n",
       " 492: 492,\n",
       " 493: 493,\n",
       " 494: 494,\n",
       " 495: 495,\n",
       " 496: 496,\n",
       " 497: 497,\n",
       " 498: 498,\n",
       " 499: 499,\n",
       " 500: 500,\n",
       " 501: 501,\n",
       " 502: 502,\n",
       " 503: 503,\n",
       " 504: 504,\n",
       " 505: 505,\n",
       " 506: 506,\n",
       " 507: 507,\n",
       " 508: 508,\n",
       " 509: 509,\n",
       " 510: 510,\n",
       " 511: 511,\n",
       " 512: 512,\n",
       " 513: 513,\n",
       " 514: 514,\n",
       " 515: 515,\n",
       " 516: 516,\n",
       " 517: 517,\n",
       " 518: 518,\n",
       " 519: 519,\n",
       " 520: 520,\n",
       " 521: 521,\n",
       " 522: 522,\n",
       " 523: 523,\n",
       " 524: 524,\n",
       " 525: 525,\n",
       " 526: 526,\n",
       " 527: 527,\n",
       " 528: 528,\n",
       " 529: 529,\n",
       " 530: 530,\n",
       " 531: 531,\n",
       " 532: 532,\n",
       " 533: 533,\n",
       " 534: 534,\n",
       " 535: 535,\n",
       " 536: 536,\n",
       " 537: 537,\n",
       " 538: 538,\n",
       " 539: 539,\n",
       " 540: 540,\n",
       " 541: 541,\n",
       " 542: 542,\n",
       " 543: 543,\n",
       " 544: 544,\n",
       " 545: 545,\n",
       " 546: 546,\n",
       " 547: 547,\n",
       " 548: 548,\n",
       " 549: 549,\n",
       " 550: 550,\n",
       " 551: 551,\n",
       " 552: 552,\n",
       " 553: 553,\n",
       " 554: 554,\n",
       " 555: 555,\n",
       " 556: 556,\n",
       " 557: 557,\n",
       " 558: 558,\n",
       " 559: 559,\n",
       " 560: 560,\n",
       " 561: 561,\n",
       " 562: 562,\n",
       " 563: 563,\n",
       " 564: 564,\n",
       " 565: 565,\n",
       " 566: 566,\n",
       " 567: 567,\n",
       " 568: 568,\n",
       " 569: 569,\n",
       " 570: 570,\n",
       " 571: 571,\n",
       " 572: 572,\n",
       " 573: 573,\n",
       " 574: 574,\n",
       " 575: 575,\n",
       " 576: 576,\n",
       " 577: 577,\n",
       " 578: 578,\n",
       " 579: 579,\n",
       " 580: 580,\n",
       " 581: 581,\n",
       " 582: 582,\n",
       " 583: 583,\n",
       " 584: 584,\n",
       " 585: 585,\n",
       " 586: 586,\n",
       " 587: 587,\n",
       " 588: 588,\n",
       " 589: 589,\n",
       " 590: 590,\n",
       " 591: 591,\n",
       " 592: 592,\n",
       " 593: 593,\n",
       " 594: 594,\n",
       " 595: 595,\n",
       " 596: 596,\n",
       " 597: 597,\n",
       " 598: 598,\n",
       " 599: 599,\n",
       " 600: 600,\n",
       " 601: 601,\n",
       " 602: 602,\n",
       " 603: 603,\n",
       " 604: 604,\n",
       " 605: 605,\n",
       " 606: 606,\n",
       " 607: 607,\n",
       " 608: 608,\n",
       " 609: 609,\n",
       " 610: 610,\n",
       " 611: 611,\n",
       " 612: 612,\n",
       " 613: 613,\n",
       " 614: 614,\n",
       " 615: 615,\n",
       " 616: 616,\n",
       " 617: 617,\n",
       " 618: 618,\n",
       " 619: 619,\n",
       " 620: 620,\n",
       " 621: 621,\n",
       " 622: 622,\n",
       " 623: 623,\n",
       " 624: 624,\n",
       " 625: 625,\n",
       " 626: 626,\n",
       " 627: 627,\n",
       " 628: 628,\n",
       " 629: 629,\n",
       " 630: 630,\n",
       " 631: 631,\n",
       " 632: 632,\n",
       " 633: 633,\n",
       " 634: 634,\n",
       " 635: 635,\n",
       " 636: 636,\n",
       " 637: 637,\n",
       " 638: 638,\n",
       " 639: 639,\n",
       " 640: 640,\n",
       " 641: 641,\n",
       " 642: 642,\n",
       " 643: 643,\n",
       " 644: 644,\n",
       " 645: 645,\n",
       " 646: 646,\n",
       " 647: 647,\n",
       " 648: 648,\n",
       " 649: 649,\n",
       " 650: 650,\n",
       " 651: 651,\n",
       " 652: 652,\n",
       " 653: 653,\n",
       " 654: 654,\n",
       " 655: 655,\n",
       " 656: 656,\n",
       " 657: 657,\n",
       " 658: 658,\n",
       " 659: 659,\n",
       " 660: 660,\n",
       " 661: 661,\n",
       " 662: 662,\n",
       " 663: 663,\n",
       " 664: 664,\n",
       " 665: 665,\n",
       " 666: 666,\n",
       " 667: 667,\n",
       " 668: 668,\n",
       " 669: 669,\n",
       " 670: 670,\n",
       " 671: 671,\n",
       " 672: 672,\n",
       " 673: 673,\n",
       " 674: 674,\n",
       " 675: 675,\n",
       " 676: 676,\n",
       " 677: 677,\n",
       " 678: 678,\n",
       " 679: 679,\n",
       " 680: 680,\n",
       " 681: 681,\n",
       " 682: 682,\n",
       " 683: 683,\n",
       " 684: 684,\n",
       " 685: 685,\n",
       " 686: 686,\n",
       " 687: 687,\n",
       " 688: 688,\n",
       " 689: 689,\n",
       " 690: 690,\n",
       " 691: 691,\n",
       " 692: 692,\n",
       " 693: 693,\n",
       " 694: 694,\n",
       " 695: 695,\n",
       " 696: 696,\n",
       " 697: 697,\n",
       " 698: 698,\n",
       " 699: 699,\n",
       " 700: 700,\n",
       " 701: 701,\n",
       " 702: 702,\n",
       " 703: 703,\n",
       " 704: 704,\n",
       " 705: 705,\n",
       " 706: 706,\n",
       " 707: 707,\n",
       " 708: 708,\n",
       " 709: 709,\n",
       " 710: 710,\n",
       " 711: 711,\n",
       " 712: 712,\n",
       " 713: 713,\n",
       " 714: 714,\n",
       " 715: 715,\n",
       " 716: 716,\n",
       " 717: 717,\n",
       " 718: 718,\n",
       " 719: 719,\n",
       " 720: 720,\n",
       " 721: 721,\n",
       " 722: 722,\n",
       " 723: 723,\n",
       " 724: 724,\n",
       " 725: 725,\n",
       " 726: 726,\n",
       " 727: 727,\n",
       " 728: 728,\n",
       " 729: 729,\n",
       " 730: 730,\n",
       " 731: 731,\n",
       " 732: 732,\n",
       " 733: 733,\n",
       " 734: 734,\n",
       " 735: 735,\n",
       " 736: 736,\n",
       " 737: 737,\n",
       " 738: 738,\n",
       " 739: 739,\n",
       " 740: 740,\n",
       " 741: 741,\n",
       " 742: 742,\n",
       " 743: 743,\n",
       " 744: 744,\n",
       " 745: 745,\n",
       " 746: 746,\n",
       " 747: 747,\n",
       " 748: 748,\n",
       " 749: 749,\n",
       " 750: 750,\n",
       " 751: 751,\n",
       " 752: 752,\n",
       " 753: 753,\n",
       " 754: 754,\n",
       " 755: 755,\n",
       " 756: 756,\n",
       " 757: 757,\n",
       " 758: 758,\n",
       " 759: 759,\n",
       " 760: 760,\n",
       " 761: 761,\n",
       " 762: 762,\n",
       " 763: 763,\n",
       " 764: 764,\n",
       " 765: 765,\n",
       " 766: 766,\n",
       " 767: 767,\n",
       " 768: 768,\n",
       " 769: 769,\n",
       " 770: 770,\n",
       " 771: 771,\n",
       " 772: 772,\n",
       " 773: 773,\n",
       " 774: 774,\n",
       " 775: 775,\n",
       " 776: 776,\n",
       " 777: 777,\n",
       " 778: 778,\n",
       " 779: 779,\n",
       " 780: 780,\n",
       " 781: 781,\n",
       " 782: 782,\n",
       " 783: 783,\n",
       " 784: 784,\n",
       " 785: 785,\n",
       " 786: 786,\n",
       " 787: 787,\n",
       " 788: 788,\n",
       " 789: 789,\n",
       " 790: 790,\n",
       " 791: 791,\n",
       " 792: 792,\n",
       " 793: 793,\n",
       " 794: 794,\n",
       " 795: 795,\n",
       " 796: 796,\n",
       " 797: 797,\n",
       " 798: 798,\n",
       " 799: 799,\n",
       " 800: 800,\n",
       " 801: 801,\n",
       " 802: 802,\n",
       " 803: 803,\n",
       " 804: 804,\n",
       " 805: 805,\n",
       " 806: 806,\n",
       " 807: 807,\n",
       " 808: 808,\n",
       " 809: 809,\n",
       " 810: 810,\n",
       " 811: 811,\n",
       " 812: 812,\n",
       " 813: 813,\n",
       " 814: 814,\n",
       " 815: 815,\n",
       " 816: 816,\n",
       " 817: 817,\n",
       " 818: 818,\n",
       " 819: 819,\n",
       " 820: 820,\n",
       " 821: 821,\n",
       " 822: 822,\n",
       " 823: 823,\n",
       " 824: 824,\n",
       " 825: 825,\n",
       " 826: 826,\n",
       " 827: 827,\n",
       " 828: 828,\n",
       " 829: 829,\n",
       " 830: 830,\n",
       " 831: 831,\n",
       " 832: 832,\n",
       " 833: 833,\n",
       " 834: 834,\n",
       " 835: 835,\n",
       " 836: 836,\n",
       " 837: 837,\n",
       " 838: 838,\n",
       " 839: 839,\n",
       " 840: 840,\n",
       " 841: 841,\n",
       " 842: 842,\n",
       " 843: 843,\n",
       " 844: 844,\n",
       " 845: 845,\n",
       " 846: 846,\n",
       " 847: 847,\n",
       " 848: 848,\n",
       " 849: 849,\n",
       " 850: 850,\n",
       " 851: 851,\n",
       " 852: 852,\n",
       " 853: 853,\n",
       " 854: 854,\n",
       " 855: 855,\n",
       " 856: 856,\n",
       " 857: 857,\n",
       " 858: 858,\n",
       " 859: 859,\n",
       " 860: 860,\n",
       " 861: 861,\n",
       " 862: 862,\n",
       " 863: 863,\n",
       " 864: 864,\n",
       " 865: 865,\n",
       " 866: 866,\n",
       " 867: 867,\n",
       " 868: 868,\n",
       " 869: 869,\n",
       " 870: 870,\n",
       " 871: 871,\n",
       " 872: 872,\n",
       " 873: 873,\n",
       " 874: 874,\n",
       " 875: 875,\n",
       " 876: 876,\n",
       " 877: 877,\n",
       " 878: 878,\n",
       " 879: 879,\n",
       " 880: 880,\n",
       " 881: 881,\n",
       " 882: 882,\n",
       " 883: 883,\n",
       " 884: 884,\n",
       " 885: 885,\n",
       " 886: 886,\n",
       " 887: 887,\n",
       " 888: 888,\n",
       " 889: 889,\n",
       " 890: 890,\n",
       " 891: 891,\n",
       " 892: 892,\n",
       " 893: 893,\n",
       " 894: 894,\n",
       " 895: 895,\n",
       " 896: 896,\n",
       " 897: 897,\n",
       " 898: 898,\n",
       " 899: 899,\n",
       " 900: 900,\n",
       " 901: 901,\n",
       " 902: 902,\n",
       " 903: 903,\n",
       " 904: 904,\n",
       " 905: 905,\n",
       " 906: 906,\n",
       " 907: 907,\n",
       " 908: 908,\n",
       " 909: 909,\n",
       " 910: 910,\n",
       " 911: 911,\n",
       " 912: 912,\n",
       " 913: 913,\n",
       " 914: 914,\n",
       " 915: 915,\n",
       " 916: 916,\n",
       " 917: 917,\n",
       " 918: 918,\n",
       " 919: 919,\n",
       " 920: 920,\n",
       " 921: 921,\n",
       " 922: 922,\n",
       " 923: 923,\n",
       " 924: 924,\n",
       " 925: 925,\n",
       " 926: 926,\n",
       " 927: 927,\n",
       " 928: 928,\n",
       " 929: 929,\n",
       " 930: 930,\n",
       " 931: 931,\n",
       " 932: 932,\n",
       " 933: 933,\n",
       " 934: 934,\n",
       " 935: 935,\n",
       " 936: 936,\n",
       " 937: 937,\n",
       " 938: 938,\n",
       " 939: 939,\n",
       " 940: 940,\n",
       " 941: 941,\n",
       " 942: 942,\n",
       " 943: 943,\n",
       " 944: 944,\n",
       " 945: 945,\n",
       " 946: 946,\n",
       " 947: 947,\n",
       " 948: 948,\n",
       " 949: 949,\n",
       " 950: 950,\n",
       " 951: 951,\n",
       " 952: 952,\n",
       " 953: 953,\n",
       " 954: 954,\n",
       " 955: 955,\n",
       " 956: 956,\n",
       " 957: 957,\n",
       " 958: 958,\n",
       " 959: 959,\n",
       " 960: 960,\n",
       " 961: 961,\n",
       " 962: 962,\n",
       " 963: 963,\n",
       " 964: 964,\n",
       " 965: 965,\n",
       " 966: 966,\n",
       " 967: 967,\n",
       " 968: 968,\n",
       " 969: 969,\n",
       " 970: 970,\n",
       " 971: 971,\n",
       " 972: 972,\n",
       " 973: 973,\n",
       " 974: 974,\n",
       " 975: 975,\n",
       " 976: 976,\n",
       " 977: 977,\n",
       " 978: 978,\n",
       " 979: 979,\n",
       " 980: 980,\n",
       " 981: 981,\n",
       " 982: 982,\n",
       " 983: 983,\n",
       " 984: 984,\n",
       " 985: 985,\n",
       " 986: 986,\n",
       " 987: 987,\n",
       " 988: 988,\n",
       " 989: 989,\n",
       " 990: 990,\n",
       " 991: 991,\n",
       " 992: 992,\n",
       " 993: 993,\n",
       " 994: 994,\n",
       " 995: 995,\n",
       " 996: 996,\n",
       " 997: 997,\n",
       " 998: 998,\n",
       " 999: 999,\n",
       " ...}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_mapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/usrvol/mappings/categorical/vocab_dict.pickle', 'rb') as f:\n",
    "    vocab_dict = pkl.load(f)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'Enzyme, 1,4-alpha-Glucan Branching',\n",
       " 1: '1 Alkyl 2 Acylphosphatidates',\n",
       " 2: '1 Carboxyglutamic Acid',\n",
       " 3: '1 Methyl 3 isobutylxanthine',\n",
       " 4: '1-Methyl-4-phenyl-1,2,3,6-tetrahydropyridine',\n",
       " 5: '1 Methyl 4 phenylpyridinium',\n",
       " 6: '1 Naphthylamine',\n",
       " 7: '1 Naphthylisothiocyanate',\n",
       " 8: 'Angiotensin II, 1-Sarcosine-8-Isoleucine',\n",
       " 9: '11 Hydroxycorticosteroids',\n",
       " 10: 'synthetic 11-hydroxycorticosteroids',\n",
       " 11: 'Hydrolase, 15-Ketosteryl Oleate',\n",
       " 12: 'RNA, 15S',\n",
       " 13: 'E2, 16,16-Dimethylprostaglandin',\n",
       " 14: 'beta-Hydroxy-5 beta-Androstan-3-One, 17',\n",
       " 15: 'Dehydrogenases, 17 beta-Hydroxysteroid',\n",
       " 16: '17-hydroxycorticosteroid',\n",
       " 17: 'Dehydrogenases, 17-Hydroxysteroid',\n",
       " 18: '17-ketosteroid',\n",
       " 19: 'Hydroxycorticosterone, 18',\n",
       " 20: 'Hydroxydesoxycorticosterone, 18',\n",
       " 21: '19 Iodocholesterol',\n",
       " 22: 'Nucleotide Phosphodiesterases, 2,3-Cyclic',\n",
       " 23: '2-5A Synthetase',\n",
       " 24: \"2' CMP\",\n",
       " 25: \"Dipyridyl, 2,2'\",\n",
       " 26: 'Acid, 2,3-Diketogulonic',\n",
       " 27: '2,4,5-trichlorophenoxyacetic acid',\n",
       " 28: '2,4 Dichlorophenoxyacetic Acid',\n",
       " 29: 'Dichloroindophenol, 2,6',\n",
       " 30: 'Mutase, 2-Acetolactate',\n",
       " 31: '2 Acetylaminofluorene',\n",
       " 32: '2 Amino 5 phosphonovalerate',\n",
       " 33: 'Aminoadipic Acid, 2',\n",
       " 34: '2 Aminopurine',\n",
       " 35: 'parlodel',\n",
       " 36: '2 bromolysergic acid diethylamide',\n",
       " 37: '2 Chloroadenosine',\n",
       " 38: '2 Hydroxy 5 nitrobenzyl Bromide',\n",
       " 39: '2 Hydroxyphenethylamine',\n",
       " 40: 'Synthase, 2-Isopropylmalate',\n",
       " 41: 'Mesnum',\n",
       " 42: '2 Naphthylamine',\n",
       " 43: 'Hydroxysteroid Dehydrogenases, 20',\n",
       " 44: '20 Methylcholanthrene',\n",
       " 45: '24,25 Dihydroxyvitamin D 3',\n",
       " 46: 'Calcidiol 1 Monooxygenase',\n",
       " 47: '25 Hydroxyvitamin D 2',\n",
       " 48: '3 alpha,20 beta Hydroxysteroid Dehydrogenase',\n",
       " 49: \"Phosphodiesterase, 3',5'-Cyclic-Nucleotide\",\n",
       " 50: \"Phosphodiesterases, 3',5'-Cyclic-GMP\",\n",
       " 51: \"3' CMP\",\n",
       " 52: \"Diaminobenzidine, 3,3'\",\n",
       " 53: \"3-3'Dichlorobenzidine\",\n",
       " 54: '3,4 Dihydroxyphenylacetic Acid',\n",
       " 55: 'DROXIDOPA',\n",
       " 56: '3,4 Methylenedioxyamphetamine',\n",
       " 57: 'beta alanine',\n",
       " 58: '3 Deazauridine',\n",
       " 59: 'meglutol',\n",
       " 60: '3 Hydroxy N methylmorphinan',\n",
       " 61: 'Dehydrogenases, 3-Hydroxyacyl CoA',\n",
       " 62: 'Acid, 3-Hydroxyanthranilic',\n",
       " 63: '3-Hydroxyproline',\n",
       " 64: 'Hydroxysteroid Dehydrogenases, 3',\n",
       " 65: 'Acid, 3-Mercaptopropionic',\n",
       " 66: '3 Methoxy 4 hydroxyphenylethanol',\n",
       " 67: '3-Oxoacyl (Acyl Carrier Protein) Synthase',\n",
       " 68: 'Docosahexaenoate',\n",
       " 69: '4-Aminobenzoic acid',\n",
       " 70: 'DALFAMPRIDINE',\n",
       " 71: '4 Butyrolactone',\n",
       " 72: '4 Epitetracycline',\n",
       " 73: '4 Fluorophenylalanine',\n",
       " 74: '4 Hydroxyaminoquinoline 1 oxide',\n",
       " 75: '4-hydroxybenzoate 3-monooxygenase',\n",
       " 76: 'OXYBATE',\n",
       " 77: '4 Hydroxycoumarins',\n",
       " 78: 'Dioxygenase, 4-Hydroxyphenylpyruvate',\n",
       " 79: '4 Nitroquinoline 1 oxide',\n",
       " 80: '5 beta Dihydrotestosterone',\n",
       " 81: '5-nucleotidase',\n",
       " 82: '5,12 diHETE',\n",
       " 83: '5,6 Dihydroxytryptamine',\n",
       " 84: '5,7 Dihydroxytryptamine',\n",
       " 85: 'acid eicosapentaenoic',\n",
       " 86: '5,8,11,14 Eicosatetraynoic Acid',\n",
       " 87: 'Androstane 3,17 diol',\n",
       " 88: \"Phosphorylase, 5-Fluoro-2'-deoxyuridine\",\n",
       " 89: '5 hydroxytryptophan',\n",
       " 90: 'Methylbufotenin',\n",
       " 91: '5 methoxytryptamine',\n",
       " 92: 'Ticlopidine hydrochloride',\n",
       " 93: '5 alpha Androstane 3 beta,17 alpha diol',\n",
       " 94: '6-Aminocaproic acid',\n",
       " 95: '6 Aminonicotinamide',\n",
       " 96: 'F1 alpha, 6-Ketoprostaglandin',\n",
       " 97: 'MERCAPTOPURINE',\n",
       " 98: '7,8-Dihydro-7,8-dihydroxybenzo(a)pyrene 9,10-oxide',\n",
       " 99: '7 Alkoxycoumarin O Dealkylase',\n",
       " 100: 'Collagen, 7S',\n",
       " 101: '7SL RNA',\n",
       " 102: '8,11,14 Eicosatrienoic Acid',\n",
       " 103: '8 Bromo Cyclic Adenosine Monophosphate',\n",
       " 104: 'OXYQUINOLINE',\n",
       " 105: 'Quinosol',\n",
       " 106: '9,10-Dimethyl-1,2-Benzanthracene',\n",
       " 107: 'Hydrase, 9,10-Epoxypalmitic Acid',\n",
       " 108: 'a fiber',\n",
       " 109: 'A gamma Globin',\n",
       " 110: 'Calcimycin',\n",
       " 111: 'A-Form DNA',\n",
       " 112: 'Temephos',\n",
       " 113: 'abattoir',\n",
       " 114: 'Abbott-46811',\n",
       " 115: 'Abbott 48999',\n",
       " 116: 'Scale, Abbreviated Injury',\n",
       " 117: 'abbreviation',\n",
       " 118: 'Abdomen+',\n",
       " 119: 'Acute Abdomen',\n",
       " 120: 'Cramps, Abdominal',\n",
       " 121: 'abdomen distended',\n",
       " 122: 'Abdominal mass nos',\n",
       " 123: 'Neoplasms, Abdominal',\n",
       " 124: 'Pain, Abdominal',\n",
       " 125: 'Muscles, Abdominal',\n",
       " 126: 'Abducens nerve structure',\n",
       " 127: 'Abelson Murine Leukemia Virus',\n",
       " 128: 'Abetalipoproteinaemia',\n",
       " 129: 'Congenital abnormal fusion, NOS',\n",
       " 130: 'Abnormal position on right side of body',\n",
       " 131: 'ABNORMAL RADIOLOGIC STUDY',\n",
       " 132: 'Body weight gain, excessive',\n",
       " 133: 'X-ray NOS abnormal',\n",
       " 134: 'Abnormality, Congenital',\n",
       " 135: 'teratologic',\n",
       " 136: 'Abnormalities of size and form of teeth',\n",
       " 137: 'Drug-Induced Abnormalities',\n",
       " 138: 'Congenital anomaly nos/multiple',\n",
       " 139: 'Radiation-Induced Abnormalities',\n",
       " 140: 'Abnormality of gastrin secretion',\n",
       " 141: 'Blood groups--ABO system',\n",
       " 142: 'Factors, ABO',\n",
       " 143: 'Abomasums',\n",
       " 144: 'Aborted Fetus (structure)',\n",
       " 145: 'Agents, Abortifacient',\n",
       " 146: 'Nonsteroidal Abortifacient Agents',\n",
       " 147: 'Steroidal Abortifacient Agents',\n",
       " 148: 'Abortions.spontaneous',\n",
       " 149: 'Applicants, Abortion',\n",
       " 150: 'Centers, Abortion',\n",
       " 151: 'Abortion, Previous',\n",
       " 152: 'law abortion',\n",
       " 153: 'Abortion on Demand (procedure)',\n",
       " 154: 'Rate, Abortion',\n",
       " 155: 'Seekers, Refused Abortion',\n",
       " 156: 'Seekers, Repeated Abortion',\n",
       " 157: 'Surveys, Abortion',\n",
       " 158: 'Criminal Abortion',\n",
       " 159: 'Drug-Induced Abortion',\n",
       " 160: 'Eugenic Abortion',\n",
       " 161: 'Habitual Abortion',\n",
       " 162: 'Abortion spontaneous incomplete NOS',\n",
       " 163: 'Legal abortion (procedure)',\n",
       " 164: 'Missed;abortion',\n",
       " 165: 'Rivanol Abortion',\n",
       " 166: 'Saline-Solution Abortion',\n",
       " 167: 'Septic Abortion',\n",
       " 168: 'Therapeutic abortion procedure, NOS',\n",
       " 169: 'threaten abortion',\n",
       " 170: 'Tubal Abortion',\n",
       " 171: 'Veterinary Abortion',\n",
       " 172: 'Abrasion or friction burn of other, multiple, and unspecified sites, without mention of infection',\n",
       " 173: 'Abreactions',\n",
       " 174: 'Abrins',\n",
       " 175: 'Abrin A',\n",
       " 176: 'Abrin C',\n",
       " 177: 'placenta; abruptio',\n",
       " 178: 'abscesses',\n",
       " 179: 'Abscisic acid',\n",
       " 180: 'agenesis',\n",
       " 181: 'ABSENT POSTERIOR TIBIAL PULSE',\n",
       " 182: 'absenteeism',\n",
       " 183: 'Absidia <fungi>',\n",
       " 184: 'Photon Absorptiometry',\n",
       " 185: 'Process of absorption',\n",
       " 186: 'Abstem',\n",
       " 187: 'abstracted',\n",
       " 188: 'Indexing and Abstracting',\n",
       " 189: 'Abu Dhabi',\n",
       " 190: 'Health Services Abuses',\n",
       " 191: 'Abutments, Dental',\n",
       " 192: 'Acacias',\n",
       " 193: 'Medical Centers, Academic',\n",
       " 194: 'Academic problems NOS',\n",
       " 195: 'Training, Academic',\n",
       " 196: 'academy',\n",
       " 197: 'Institutes and Academies',\n",
       " 198: 'Acanthameba',\n",
       " 199: 'Keratitis, Acanthamoeba',\n",
       " 200: 'Acanthocephala',\n",
       " 201: 'acanthrocyte',\n",
       " 202: 'Acantholyses',\n",
       " 203: 'Nigricans, Acanthosis',\n",
       " 204: 'ORDER ACARINA',\n",
       " 205: 'Acarus, NOS',\n",
       " 206: 'Accelerated particle, NOS',\n",
       " 207: 'accelerations',\n",
       " 208: 'Accelographies',\n",
       " 209: 'Health Care Acceptability',\n",
       " 210: 'Processes, Acceptance',\n",
       " 211: 'acceptor',\n",
       " 212: 'Accessory nerve structure',\n",
       " 213: 'Accident caused by caustic and corrosive substances',\n",
       " 214: 'Accident caused by excessive heat due to weather conditions',\n",
       " 215: 'Accident caused by unspecified fire',\n",
       " 216: 'Accident due to changes in air pressure due to diving',\n",
       " 217: 'Accident caused by lightning',\n",
       " 218: 'Accident due to residence or prolonged visit at high altitude',\n",
       " 219: 'Prevention, Accident',\n",
       " 220: 'Proneness, Accident',\n",
       " 221: 'Falls, Accidental',\n",
       " 222: 'Accidental mechanical suffocation by unspecified means',\n",
       " 223: 'Accidental poisoning from other fish',\n",
       " 224: 'Accident NOS',\n",
       " 225: 'Incised wound, NOS',\n",
       " 226: 'Accidents due to excessive heat of unspecified origin',\n",
       " 227: 'Aircraft accident',\n",
       " 228: 'Accidents, Home',\n",
       " 229: 'Accidents, Industrial',\n",
       " 230: 'Accident, Occupational',\n",
       " 231: 'traffic accident',\n",
       " 232: 'Acclimatisation',\n",
       " 233: 'ACCOMMODATION ERROR',\n",
       " 234: 'Visual accommodation, NOS',\n",
       " 235: 'ACCOUNTANTS',\n",
       " 236: 'Accounting (activity)',\n",
       " 237: 'Accounts Payable',\n",
       " 238: 'Accounts Payable and Receivable',\n",
       " 239: 'Accreditations',\n",
       " 240: 'acculturation',\n",
       " 241: '(±)-acebutolol',\n",
       " 242: 'aceclidin',\n",
       " 243: 'Acedapsone',\n",
       " 244: 'Acedoxin',\n",
       " 245: 'Acenaphthenes',\n",
       " 246: 'Acénocoumarol',\n",
       " 247: 'ACEPROMAZINE',\n",
       " 248: 'Acepromazine maleate',\n",
       " 249: 'Acetabularia',\n",
       " 250: 'Bone structure of acetabulum',\n",
       " 251: 'Acetaco',\n",
       " 252: 'ACETALDEHYDE',\n",
       " 253: 'acetal',\n",
       " 254: 'Acetamides',\n",
       " 255: 'ACETAMINOPHEN',\n",
       " 256: 'acetanilid',\n",
       " 257: 'Acetanilides',\n",
       " 258: 'ACETATE',\n",
       " 259: 'Kinase, Acetate',\n",
       " 260: 'acetates',\n",
       " 261: 'Acétazolamide',\n",
       " 262: 'acetic acid,',\n",
       " 263: 'Acids, Acetic',\n",
       " 264: 'Acetic Anhydride',\n",
       " 265: 'Anhydrides, Acetic',\n",
       " 266: 'Acetoacetate',\n",
       " 267: 'Acetobacter',\n",
       " 268: 'Acétohexamide',\n",
       " 269: 'Synthetase I, Acetohydroxyacid',\n",
       " 270: 'acetoin',\n",
       " 271: 'Dehydrogenase, Acetoin',\n",
       " 272: 'Synthase, Acetolactate',\n",
       " 273: 'acetones',\n",
       " 274: 'Bodies, Acetone',\n",
       " 275: 'Acetonitriles',\n",
       " 276: 'Acetophenones',\n",
       " 277: 'Acetoxyacetylaminofluorene',\n",
       " 278: 'Synthase III, Acetoxyhydroxyacid',\n",
       " 279: 'Acid, Acetrizoic',\n",
       " 280: 'C-Acetyltransferase, Acetyl-CoA',\n",
       " 281: 'C-Acyltransferase, Acetyl-CoA',\n",
       " 282: 'Carboxylase, Acetyl-CoA',\n",
       " 283: 'Hydrolase, Acetyl-CoA',\n",
       " 284: 'ACETYL COENZYME A',\n",
       " 285: 'Ligase, Acetate-CoA',\n",
       " 286: 'acetylaminofluorene',\n",
       " 287: 'beta-Casein Acetylated, Dephosphorylated',\n",
       " 288: 'Acetylations',\n",
       " 289: 'ACETYLCARNITINE',\n",
       " 290: 'ACETYLCHOLINE',\n",
       " 291: 'acetylcholinesterases',\n",
       " 292: 'Inhibitors, Acetylcholinesterase',\n",
       " 293: 'ACETYLCYSTEINE',\n",
       " 294: 'Acetyldigitoxins',\n",
       " 295: 'acetyldigoxin',\n",
       " 296: 'Acetylene',\n",
       " 297: 'Acetylesterase',\n",
       " 298: 'acetylgalactosamine',\n",
       " 299: 'Acetylglucosamine',\n",
       " 300: 'beta-N-Acetylglucosaminidase',\n",
       " 301: 'Isoglutamine, Acetylmuramyl Alanyl',\n",
       " 302: 'acecainide',\n",
       " 303: 'Acetylserotonin N-Methyltransferase',\n",
       " 304: 'Acetylthiocholine',\n",
       " 305: 'Acetylthiocholinesterase',\n",
       " 306: 'acetyltransferases',\n",
       " 307: 'achievements',\n",
       " 308: 'Structure of Achilles tendon',\n",
       " 309: 'achlorhydria',\n",
       " 310: 'Acholeplasma, NOS',\n",
       " 311: 'Acholeplasma laidlawii',\n",
       " 312: 'Acholeplasmataceae',\n",
       " 313: 'Achondrogenesis, NOS',\n",
       " 314: 'Achondroplasias',\n",
       " 315: 'Achromatiaceae',\n",
       " 316: 'Achromatium',\n",
       " 317: 'Achromia parasitica',\n",
       " 318: 'Achromobacter, NOS',\n",
       " 319: 'Achromobacteriaceae',\n",
       " 320: 'Topicycline',\n",
       " 321: 'Acid Cysteine Proteinase Inhibitor',\n",
       " 322: 'DNase, Acid',\n",
       " 323: 'Etching, Dental Acid',\n",
       " 324: 'Lipase, Acid',\n",
       " 325: 'Lipase A, Acid',\n",
       " 326: 'Lipase B, Acid',\n",
       " 327: 'Lipase I, Acid',\n",
       " 328: 'Lipase II, Acid',\n",
       " 329: 'acid phosphatase',\n",
       " 330: 'Rain, Acid',\n",
       " 331: 'Ribonuclease, Acid',\n",
       " 332: 'Equilibrium, Acid-Base',\n",
       " 333: 'Imbalance, Acid-Base',\n",
       " 334: 'Acidin Pepsin',\n",
       " 335: 'acidoses',\n",
       " 336: 'Lactic Acidosis',\n",
       " 337: 'Renal acidosis tubular',\n",
       " 338: 'Respiratory Acidosis',\n",
       " 339: 'ACID',\n",
       " 340: 'Acyclic Acids',\n",
       " 341: 'Aldehydic Acids',\n",
       " 342: 'Carbocyclic Acids',\n",
       " 343: 'Acids, Heterocyclic',\n",
       " 344: 'Noncarboxylic Acids',\n",
       " 345: 'acidulated fluoride phosphate',\n",
       " 346: 'Acignost',\n",
       " 347: 'Acinetobacter, NOS',\n",
       " 348: 'Infections, Acinetobacter',\n",
       " 349: 'Acladiosis',\n",
       " 350: 'Aclarubicin',\n",
       " 351: 'acne; vulgaris',\n",
       " 352: 'Keloid, Acne',\n",
       " 353: 'Acomys',\n",
       " 354: 'Hydratase, Aconitate',\n",
       " 355: 'Aconites',\n",
       " 356: 'Aconitic acid',\n",
       " 357: 'ACONITINE',\n",
       " 358: 'Tests, Acoustic Impedance',\n",
       " 359: 'Nerve (Acoustic)',\n",
       " 360: 'Vestibulocochlear Nerve Disease',\n",
       " 361: 'Stimulation, Acoustic',\n",
       " 362: 'acoustic',\n",
       " 363: 'Complete obstruction',\n",
       " 364: 'Acquired coagulation factor deficiency, NOS',\n",
       " 365: 'Acquired deformities of auricle or pinna',\n",
       " 366: 'Acquired chest and rib deformity',\n",
       " 367: 'Acquired obstruction of pylorus',\n",
       " 368: 'Syndrome, Acquired Immunodeficiency',\n",
       " 369: 'Acremonium, NOS',\n",
       " 370: 'Orange, Acridine',\n",
       " 371: 'acridine',\n",
       " 372: 'ACRIFLAVINE',\n",
       " 373: \"apert's syndrome\",\n",
       " 374: 'Acrodermatitis, NOS',\n",
       " 375: 'acrodynia',\n",
       " 376: 'acrokeratosis',\n",
       " 377: 'Acrolactine',\n",
       " 378: 'acrolein',\n",
       " 379: 'acromegaly',\n",
       " 380: 'Hypersomatotropic gigantism',\n",
       " 381: 'Acromioclavicular joint structure',\n",
       " 382: 'Acromions',\n",
       " 383: 'Acronine',\n",
       " 384: 'Akrosin',\n",
       " 385: 'Acrosomes',\n",
       " 386: 'acrylamides',\n",
       " 387: 'acrylates',\n",
       " 388: 'Resins, Acrylic',\n",
       " 389: 'acrylonitrile',\n",
       " 390: 'Syndrome, Ectopic ACTH',\n",
       " 391: 'Actihaemyl',\n",
       " 392: 'Protein, Actin-Binding',\n",
       " 393: 'acting out',\n",
       " 394: 'Actinin',\n",
       " 395: 'Actinium, NOS',\n",
       " 396: 'Actinobacillosis, NOS',\n",
       " 397: 'Actinobacillus, NOS',\n",
       " 398: 'Infections, Actinobacillus',\n",
       " 399: 'Actinomyces, NOS',\n",
       " 400: 'Actinomyces israelii (bacteria)',\n",
       " 401: 'Actinomycetaceae',\n",
       " 402: 'Actinomycetales',\n",
       " 403: 'Infections, Actinomycetales',\n",
       " 404: 'Actinomycins',\n",
       " 405: 'actinomycoses',\n",
       " 406: 'abdominal actinomycosis',\n",
       " 407: 'cervicofacial actinomycosis',\n",
       " 408: 'Actinomycotic infection of other specified sites',\n",
       " 409: 'SPECTINOMYCIN',\n",
       " 410: 'actin',\n",
       " 411: 'Potentials, Action',\n",
       " 412: 'Research, Action',\n",
       " 413: 'charcoal activated',\n",
       " 414: 'Armed macrophage',\n",
       " 415: 'Analysis, Activation',\n",
       " 416: 'Appliances, Activator',\n",
       " 417: 'Active Euthanasia',\n",
       " 418: 'Activities of Daily Living (activity)',\n",
       " 419: 'Cycles, Activity',\n",
       " 420: 'Meromyosin, Actoheavy',\n",
       " 421: 'Actomyosins',\n",
       " 422: 'Subfragments, Actomyosin',\n",
       " 423: 'Analysis, Actuarial',\n",
       " 424: 'Acupuncture therapy discipline',\n",
       " 425: 'Analgesia, Acupuncture',\n",
       " 426: 'Anesthesia, Acupuncture',\n",
       " 427: 'Points, Acupuncture',\n",
       " 428: 'Acute abscess',\n",
       " 429: 'Acute adenoviral follicular conjunctivitis',\n",
       " 430: 'alcoholic liver disease acute',\n",
       " 431: 'Acute and subacute liver necrosis (disorder)',\n",
       " 432: 'acute atopic conjunctivitis',\n",
       " 433: 'Acute bronchiolitis, NOS',\n",
       " 434: 'Disease, Acute',\n",
       " 435: 'Acute erythraemia and erythroleukaemia',\n",
       " 436: 'Laryngitis acute NOS',\n",
       " 437: 'Acute Necrotizing Encephalitis, Herpetic',\n",
       " 438: 'Acute Pancreatitis',\n",
       " 439: 'acute periodontitis',\n",
       " 440: 'pharyngitis nos acute',\n",
       " 441: 'alpha 1-Protein, Acute-Phase',\n",
       " 442: 'Glycoproteins, Acute-Phase',\n",
       " 443: 'Proteins, Acute-Phase',\n",
       " 444: 'Reaction, Acute-Phase',\n",
       " 445: 'acute; thyroiditis',\n",
       " 446: 'Acute tonsillitis, NOS',\n",
       " 447: 'Acute vascular insufficiency of intestine (disorder)',\n",
       " 448: 'Hepatic Necrosis, Massive',\n",
       " 449: 'Acute ill-defined cerebrovascular disease, NOS',\n",
       " 450: 'aciclovir',\n",
       " 451: 'Protein, Acyl Carrier',\n",
       " 452: 'Coenzyme A, Acyl',\n",
       " 453: 'acylation',\n",
       " 454: 'Acylglycerol',\n",
       " 455: 'Cytidylyltransferase, Acylneuraminate',\n",
       " 456: 'acyltransferases',\n",
       " 457: 'adamantane',\n",
       " 458: 'Syndrome Stokes-Adams',\n",
       " 459: 'Adaptation, Biological',\n",
       " 460: 'Ocular Adaptation',\n",
       " 461: 'Physiological Adaptation',\n",
       " 462: 'Disease, Addison',\n",
       " 463: 'aden',\n",
       " 464: 'ADENINE',\n",
       " 465: 'Mitochondrial ADP, ATP Translocases',\n",
       " 466: 'Nucleotides, Adenine',\n",
       " 467: 'Phosphoribosyltransferase, Adenine',\n",
       " 468: 'Adenitis, NOS',\n",
       " 469: 'dependoviruses',\n",
       " 470: 'adenocarcinomas',\n",
       " 471: 'Papillary adenocarcinoma NOS',\n",
       " 472: 'Adenofibroma NOS',\n",
       " 473: 'Adenoidectomy, NOS',\n",
       " 474: 'Adenoiditis, NOS',\n",
       " 475: 'Adenoidal structure',\n",
       " 476: 'Adenolymphomas',\n",
       " 477: 'ADENOMA NOS',\n",
       " 478: 'Basophil Adenoma',\n",
       " 479: 'Chromophobe Adenoma',\n",
       " 480: 'Acidophil Adenoma',\n",
       " 481: 'Adenosarcomas',\n",
       " 482: 'Adénosine',\n",
       " 483: \"Adenosine 2' Phosphate\",\n",
       " 484: \"Adenosine 3' Phosphate\",\n",
       " 485: 'Phosphaden',\n",
       " 486: 'Receptor, Adenosine A1',\n",
       " 487: 'Receptors, Adenosine A2',\n",
       " 488: 'cyclic amp',\n",
       " 489: 'adenosine deaminase',\n",
       " 490: 'Diphosphate, Adenosine',\n",
       " 491: 'Glucose, Adenosine Diphosphate',\n",
       " 492: 'Ribose, Adenosine Diphosphate',\n",
       " 493: 'Sugars, Adenosine Diphosphate',\n",
       " 494: 'Kinase, Adenosine',\n",
       " 495: 'Adenosine monophosphate',\n",
       " 496: 'Phosphosulfate, Adenosine',\n",
       " 497: 'P1 Receptors, Purinergic',\n",
       " 498: 'adenosine triphosphatase',\n",
       " 499: 'ATPase, F1',\n",
       " 500: 'Ca^2+^-transporting ATPase',\n",
       " 501: 'Ca(2+) Mg(2+)-ATPase',\n",
       " 502: 'Adenosinetriphosphatase, Actomyosin',\n",
       " 503: 'Na(+)-K(+)-Exchanging ATPase',\n",
       " 504: 'ADENOSINE TRIPHOSPHATE',\n",
       " 505: 'Adenovirus, NOS',\n",
       " 506: 'Adenovirus Core Protein VII',\n",
       " 507: 'Adenovirus infection in conditions classified elsewhere and of unspecified site',\n",
       " 508: 'Infections, Adenovirus',\n",
       " 509: 'Infections, Human Adenovirus',\n",
       " 510: 'Aviadenovirus, NOS',\n",
       " 511: 'Human Adenoviruses',\n",
       " 512: 'Simian Adenoviruses',\n",
       " 513: 'Cyclase, Adenylate',\n",
       " 514: 'adenylate kinase',\n",
       " 515: 'Adenylosuccinate lyase',\n",
       " 516: 'Synthase, Adenylosuccinate',\n",
       " 517: 'Imidodiphosphate, Adenylyl',\n",
       " 518: 'Adhesions NOS postoperative',\n",
       " 519: 'Adhesions, Tissue',\n",
       " 520: 'adhesive bandages',\n",
       " 521: 'Adhesive tape, device',\n",
       " 522: 'Adhesivenesses',\n",
       " 523: 'adhesive',\n",
       " 524: 'Adicin',\n",
       " 525: 'Syndrome, Adie',\n",
       " 526: 'Adipex P',\n",
       " 527: 'Adiphenine',\n",
       " 528: 'Acids, Adipic',\n",
       " 529: 'Adipocere',\n",
       " 530: 'adipose tissues',\n",
       " 531: 'adiposis; dolorosa',\n",
       " 532: 'Adjustment disorders, With depressed mood',\n",
       " 533: 'Adjustment disorders, With disturbance of conduct',\n",
       " 534: 'Adjustment disorders, With mixed disturbance of emotions and conduct',\n",
       " 535: 'adjustment disorder with mixed emotional features',\n",
       " 536: 'adjustment disorder with physical complaints',\n",
       " 537: 'ADJUSTMENT DISORDER WITHDRAWAL',\n",
       " 538: 'adjustment disorder with work or academic inhibition',\n",
       " 539: 'Disorders, Adjustment',\n",
       " 540: 'Brief depressive adjustment reaction',\n",
       " 541: 'Discretionary Adjustment',\n",
       " 542: 'Adjuvants, Immunologic',\n",
       " 543: 'Adjuvants, Pharmaceutical',\n",
       " 544: 'Administration occupational activities',\n",
       " 545: 'administration and dosage',\n",
       " 546: 'Buccal Administration',\n",
       " 547: 'Cutaneous Administration',\n",
       " 548: 'Drug Administration, Inhalation',\n",
       " 549: 'Intranasal Administration',\n",
       " 550: 'Intravaginal Administration',\n",
       " 551: 'Intravesical Administration',\n",
       " 552: 'Oral Administration',\n",
       " 553: 'Sublingual Route of Drug Administration',\n",
       " 554: 'Topical Administration',\n",
       " 555: 'Administrative District',\n",
       " 556: 'Personnel, Administrative',\n",
       " 557: 'Tests, Routine Admission',\n",
       " 558: 'Department, Admitting',\n",
       " 559: 'Hospital Admitting Department',\n",
       " 560: 'Uterine adnexae structure',\n",
       " 561: 'Diseases, Adnexal',\n",
       " 562: 'adnexitis',\n",
       " 563: 'adolescence',\n",
       " 564: 'Behavior, Adolescent',\n",
       " 565: 'Medicine, Adolescent',\n",
       " 566: 'Psychiatry, Adolescent',\n",
       " 567: 'Adolescent Psychology',\n",
       " 568: 'Hospitalized Adolescent',\n",
       " 569: 'Institutionalized Adolescent',\n",
       " 570: 'Female Adolescents',\n",
       " 571: 'Male Adolescents',\n",
       " 572: 'adona',\n",
       " 573: 'adoptions',\n",
       " 574: 'Adosterol',\n",
       " 575: 'ADP Receptors',\n",
       " 576: 'Cortex, Adrenal',\n",
       " 577: 'Diseases, Adrenal Cortex',\n",
       " 578: 'adrenal cortex function test',\n",
       " 579: 'Hormones, Adrenal Cortex',\n",
       " 580: 'Neoplasms, Adrenal Cortex',\n",
       " 581: 'Gland Diseases, Adrenal',\n",
       " 582: 'Hyperfunction, Adrenal Gland',\n",
       " 583: 'Adrenal Gland Hypofunction',\n",
       " 584: 'Neoplasms, Adrenal Gland',\n",
       " 585: 'Glands, Adrenal',\n",
       " 586: 'Congenital adrenal hyperplasia, NOS',\n",
       " 587: 'Medulla, Adrenal',\n",
       " 588: 'Tumor, Adrenal Rest',\n",
       " 589: 'Adrenalectomy NOS',\n",
       " 590: 'EPINEPHRINE HYDROCHLORIDE',\n",
       " 591: 'Agents, Adrenergic',\n",
       " 592: 'Adrenergic alpha 1 Receptors',\n",
       " 593: 'Adrenergic alpha 2 Receptors',\n",
       " 594: 'alpha-Agonists, Adrenergic',\n",
       " 595: 'alpha Adrenergic Antagonists',\n",
       " 596: 'Adrenergic beta 1 Receptors',\n",
       " 597: 'Adrenergic beta 2 Receptors',\n",
       " 598: 'beta-Agonists, Adrenergic',\n",
       " 599: 'beta Adrenergic Antagonists',\n",
       " 600: 'Fibers, Adrenergic',\n",
       " 601: 'Neuron Agents, Adrenergic',\n",
       " 602: 'Agonists, Adrenergic',\n",
       " 603: 'Adrenochrome',\n",
       " 604: 'corticotrophin',\n",
       " 605: 'Adrenodoxins',\n",
       " 606: 'Adsorption (natural process)',\n",
       " 607: 'adults',\n",
       " 608: 'Adult antisocial behaviour',\n",
       " 609: 'Adult maltreatment syndrome',\n",
       " 610: 'Directives, Advance',\n",
       " 611: 'ADVANCED TROPHIC CHANGES OF THE FOOT',\n",
       " 612: 'aspects of adverse effects',\n",
       " 613: 'Advertising (activity)',\n",
       " 614: 'Advisory Services',\n",
       " 615: 'Genus: Aedes',\n",
       " 616: 'Aequamen',\n",
       " 617: 'Aequorins',\n",
       " 618: 'beta-Lactamase, AER-I',\n",
       " 619: 'aeroallergens',\n",
       " 620: 'KLEBSIELLA PNEUMONIAE',\n",
       " 621: 'Aerobic Exercise',\n",
       " 622: 'Aerobioses',\n",
       " 623: 'Aerococcus, NOS',\n",
       " 624: 'Aeromonas, NOS',\n",
       " 625: 'aerophagy',\n",
       " 626: 'Propellants, Aerosol',\n",
       " 627: 'Aerosol',\n",
       " 628: 'Medicine, Aerospace',\n",
       " 629: 'Polymyxin B sulphate',\n",
       " 630: 'Dihydrobromide, beta-Aminoethylisothiuronium',\n",
       " 631: 'Trittico',\n",
       " 632: 'Affect (mental function)',\n",
       " 633: 'Psychotic Affective Disorders',\n",
       " 634: 'Symptoms, Affective',\n",
       " 635: 'Syndrome, Afferent Loop',\n",
       " 636: 'Pathways, Afferent',\n",
       " 637: 'Labels, Affinity',\n",
       " 638: 'afghanistan',\n",
       " 639: 'afibrinogenaemia',\n",
       " 640: 'Aflatoxins, NOS',\n",
       " 641: 'Africa',\n",
       " 642: 'Africa South of the Sahara',\n",
       " 643: 'Central Africa',\n",
       " 644: 'Eastern Africa',\n",
       " 645: 'Northern Africa',\n",
       " 646: 'Southern Africa',\n",
       " 647: 'Western Africa',\n",
       " 648: 'Sickness, African Horse',\n",
       " 649: 'African horse sickness virus, NOS',\n",
       " 650: 'African Swine Fever',\n",
       " 651: 'African swine fever virus',\n",
       " 652: 'Afrikaans Language',\n",
       " 653: 'aftercare',\n",
       " 654: 'afterimages',\n",
       " 655: 'Agalega Islands',\n",
       " 656: 'Agammaglobulinemia NOS',\n",
       " 657: 'agars',\n",
       " 658: 'Dilution Count, Agar',\n",
       " 659: 'Agaricaceae',\n",
       " 660: 'Agaricales',\n",
       " 661: 'Agaricus, NOS',\n",
       " 662: 'Age, NOS',\n",
       " 663: 'Age Determination by Skeleton',\n",
       " 664: 'Age Determination by Teeth',\n",
       " 665: 'Distribution, Age',\n",
       " 666: 'Factors, Age',\n",
       " 667: 'Age Reporting',\n",
       " 668: 'Age-Related Osteoporosis',\n",
       " 669: 'Birth Rate, Age-Specific',\n",
       " 670: 'Rate, Age-Specific Death',\n",
       " 671: 'Rate, Age-Specific Fertility',\n",
       " 672: 'Elderly (population group)',\n",
       " 673: 'aged over 80',\n",
       " 674: 'United States. Agency for International Development',\n",
       " 675: 'agglutinations',\n",
       " 676: 'Tests, Agglutination',\n",
       " 677: 'Agglutinin, NOS',\n",
       " 678: 'Agglutinogen 2, Bordetella Pertussis',\n",
       " 679: 'Aggregation Bias',\n",
       " 680: 'Aggressive Behaviour',\n",
       " 681: 'ageing',\n",
       " 682: 'Agmatine',\n",
       " 683: 'Myelofibrosis, Primary',\n",
       " 684: 'Agnosias',\n",
       " 685: 'Behavior, Agonistic',\n",
       " 686: 'agoraphobias',\n",
       " 687: 'Agoraphobia without history of panic disorder, NOS',\n",
       " 688: 'Agovirin',\n",
       " 689: 'agranulocytoses',\n",
       " 690: 'Agraphia NOS',\n",
       " 691: 'Agricultural Development (occupation)',\n",
       " 692: \"Workers' Diseases, Agricultural\",\n",
       " 693: 'agriculture',\n",
       " 694: 'Agrobacterium, NOS',\n",
       " 695: 'Aid to Families with Dependent Children',\n",
       " 696: 'Aid to Visually Impaired Persons',\n",
       " 697: 'Aid to Totally Disabled Persons',\n",
       " 698: 'Dementia Complex, AIDS',\n",
       " 699: 'Serology, AIDS',\n",
       " 700: 'AIDS-related complex',\n",
       " 701: 'AINHUM',\n",
       " 702: 'Air (substance)',\n",
       " 703: 'air bubbles',\n",
       " 704: 'Conditioning, Air',\n",
       " 705: 'Ionization, Air',\n",
       " 706: 'Microbiology, Air',\n",
       " 707: 'Movements, Air',\n",
       " 708: 'Pollutants, Air',\n",
       " 709: 'Pollutants, Air Environmental',\n",
       " 710: 'Pollutants, Occupational Air',\n",
       " 711: 'Radioactive Air Pollutants',\n",
       " 712: 'Pollution, Air',\n",
       " 713: 'Radioactive Air Pollution',\n",
       " 714: 'Pressure, Air',\n",
       " 715: 'Sacs, Air',\n",
       " 716: 'aircrafts',\n",
       " 717: 'airsickness',\n",
       " 718: 'Obstruction, Airway',\n",
       " 719: 'Airway Resistance Test',\n",
       " 720: '(+)-Ajmaline',\n",
       " 721: 'Mutism, Akinetic',\n",
       " 722: 'Petit Mal, Akinetic',\n",
       " 723: 'AKR murine leukemia virus',\n",
       " 724: 'alabama',\n",
       " 725: 'alabaster',\n",
       " 726: 'Alamethicin',\n",
       " 727: 'Alanines',\n",
       " 728: 'Transaminase, Alanine',\n",
       " 729: 'Racemase, Alanine',\n",
       " 730: 'Ligase, Alanine-tRNA',\n",
       " 731: 'alaska',\n",
       " 732: 'Minor, Variola',\n",
       " 733: 'albania',\n",
       " 734: 'Albanian Language',\n",
       " 735: 'ALBENDAZOLE',\n",
       " 736: 'Alberta province',\n",
       " 737: 'Albinism, NOS',\n",
       " 738: 'Albumin, abnormal presence',\n",
       " 739: 'Albumen',\n",
       " 740: 'Albuminuria NOS',\n",
       " 741: 'ALBUTEROL',\n",
       " 742: 'Alcaligenes, NOS',\n",
       " 743: 'Alchemy',\n",
       " 744: 'Blue, Alcian',\n",
       " 745: 'Alclofénac',\n",
       " 746: 'Amnestic Disorder, Alcohol',\n",
       " 747: 'alcohol dehydrogenases',\n",
       " 748: 'Alcohol Dehydrogenase I',\n",
       " 749: 'Alcohol Dehydrogenase II',\n",
       " 750: 'Deterrents, Alcohol',\n",
       " 751: 'Alcohol consumption NOS',\n",
       " 752: 'idiosyncratic alcohol intoxication',\n",
       " 753: 'Oxidoreductases, Alcohol',\n",
       " 754: 'Use Disorders, Alcohol',\n",
       " 755: 'Delirium, Alcohol Withdrawal',\n",
       " 756: 'éthanol',\n",
       " 757: 'Methanol',\n",
       " 758: '1-Propanol',\n",
       " 759: 'Beverages (Alcoholic)',\n",
       " 760: 'Intoxication, Alcoholic',\n",
       " 761: 'Silver Nitrate, Alcoholic',\n",
       " 762: 'Anonymous, Alcoholics',\n",
       " 763: 'Intoxication, Chronic Alcoholic',\n",
       " 764: 'ALCOHOL',\n",
       " 765: 'Pentanols',\n",
       " 766: 'butanol',\n",
       " 767: 'hexanol',\n",
       " 768: 'octanol',\n",
       " 769: 'Alcuronium',\n",
       " 770: 'Aldehyde dehydrogenase (NAD^+^)',\n",
       " 771: 'Aldehyde Dehydrogenase E1',\n",
       " 772: 'Aldehyde Dehydrogenase E2',\n",
       " 773: 'Oxidoreductases, Aldehyde',\n",
       " 774: 'Aldehyde-lyase',\n",
       " 775: 'Aldehyde, NOS',\n",
       " 776: 'Alderney Island',\n",
       " 777: 'ALDICARB',\n",
       " 778: 'Aldolase A',\n",
       " 779: 'Aldolase B',\n",
       " 780: 'Aldolase C',\n",
       " 781: 'Reductase, Aldehyde',\n",
       " 782: 'Reductase Ia, Aldose',\n",
       " 783: 'Reductase Ib, Aldose',\n",
       " 784: '(+)-aldosterone',\n",
       " 785: 'Antagonists, Aldosterone',\n",
       " 786: 'Aldridge Procedure',\n",
       " 787: 'aldrin',\n",
       " 788: 'Mink Disease, Aleutian',\n",
       " 789: 'Aleutian mink disease virus',\n",
       " 790: 'alexias',\n",
       " 791: 'Alexia and dyslexia',\n",
       " 792: 'Alexithymias',\n",
       " 793: 'alfalfa',\n",
       " 794: 'Alfenta',\n",
       " 795: 'alfentanyl',\n",
       " 796: 'Alfeprol',\n",
       " 797: 'alga',\n",
       " 798: 'Algae, Brown (eukaryote)',\n",
       " 799: 'chlorophyta',\n",
       " 800: 'Rhodophyta',\n",
       " 801: 'Bloom, Algal',\n",
       " 802: 'algeria',\n",
       " 803: 'Acetophenide, Algestone',\n",
       " 804: 'alginates',\n",
       " 805: 'Algor mortis',\n",
       " 806: 'algorithms',\n",
       " 807: 'Algotropyl',\n",
       " 808: 'alien',\n",
       " 809: 'Aliphatic Acid',\n",
       " 810: 'Alkadienes',\n",
       " 811: 'Alkalescens-Dispar Group',\n",
       " 812: 'alkalis',\n",
       " 813: 'DNase, Alkaline',\n",
       " 814: 'Hematin D-575, Alkaline',\n",
       " 815: 'Phosphatase, Alkaline',\n",
       " 816: 'Ribonuclease, Alkaline',\n",
       " 817: 'Alkaloid, NOS',\n",
       " 818: 'alkaloses',\n",
       " 819: 'Respiratory Alkalosis',\n",
       " 820: 'Alkane',\n",
       " 821: 'alcaptonuria',\n",
       " 822: 'Alkene',\n",
       " 823: 'Phospholipids, Alkenyl Ether',\n",
       " 824: 'Glyceryl Ethers, Alkyl',\n",
       " 825: 'Alkanesulfonates',\n",
       " 826: 'Agents, Alkylating',\n",
       " 827: 'Alkylations',\n",
       " 828: 'Compounds, Alkylmercury',\n",
       " 829: 'Alkyne',\n",
       " 830: 'Vehicles, All Terrain',\n",
       " 831: 'ALLANTOIN',\n",
       " 832: 'Structure of allantois',\n",
       " 833: 'allele',\n",
       " 834: 'ALLERGEN',\n",
       " 835: 'Hypersensitivity granuloma',\n",
       " 836: 'Atopic Rhinitis',\n",
       " 837: 'Allergic rhinitis due to other allergens',\n",
       " 838: 'allergic; pneumonia',\n",
       " 839: 'Specialty, Allergy',\n",
       " 840: 'Immunology and Allergy',\n",
       " 841: 'ALLERGY TO DIAGNOSTIC AGENT',\n",
       " 842: 'allethrins',\n",
       " 843: 'Occupations, Allied Health',\n",
       " 844: 'Personnel, Allied Health',\n",
       " 845: 'alligator',\n",
       " 846: 'Crocodiles and Alligators',\n",
       " 847: 'Allium, NOS',\n",
       " 848: 'Allouridine',\n",
       " 849: 'Alloantigen',\n",
       " 850: 'Allobarbital',\n",
       " 851: 'Allogenic disease',\n",
       " 852: 'Alloisoleucine',\n",
       " 853: 'Allomyces',\n",
       " 854: 'Hydrolase, Allophanate',\n",
       " 855: 'ALLOPURINOL',\n",
       " 856: 'Regulation, Allosteric',\n",
       " 857: 'Site, Allosteric',\n",
       " 858: 'ALLOXAN',\n",
       " 859: 'Diabetes, Alloxan',\n",
       " 860: 'Alloy',\n",
       " 861: 'Compounds, Allyl',\n",
       " 862: 'allylamines',\n",
       " 863: 'Allyloestrenol',\n",
       " 864: 'Allylglycine',\n",
       " 865: 'Allylisopropylacetamide',\n",
       " 866: 'Almitrine',\n",
       " 867: 'Almitrine Dimesylate',\n",
       " 868: 'aloes',\n",
       " 869: 'alopecias',\n",
       " 870: 'alopecia; areata',\n",
       " 871: 'Follicular Mucinosis',\n",
       " 872: 'Alopecia, drug',\n",
       " 873: 'Alopecia, endocrine',\n",
       " 874: 'Alopecia, metabolic',\n",
       " 875: 'Nutritional alopecia, NOS',\n",
       " 876: 'Hypotrichosis following infectious disease',\n",
       " 877: 'postpartum; alopecia',\n",
       " 878: 'Alopecia syphilitic',\n",
       " 879: 'traumatic alopecia',\n",
       " 880: 'Alouatta, NOS',\n",
       " 881: 'Aloxiprimum',\n",
       " 882: 'Vicugna pacos',\n",
       " 883: 'alpha 1 Acute Phase Globulin',\n",
       " 884: 'alpha(1)-Antichymotrypsin',\n",
       " 885: '1 alpha antitrypsin',\n",
       " 886: 'alpha 1 Macroglobulin',\n",
       " 887: 'alpha 2 Acute Phase Globulins',\n",
       " 888: 'alpha 2-Glucoprotein',\n",
       " 889: 'alpha-Interferon',\n",
       " 890: 'alpha Crystallin A Chain',\n",
       " 891: 'alpha Atrial Natriuretic Peptide',\n",
       " 892: 'Alpha-fetoprotein',\n",
       " 893: 'alpha-Globulin, NOS',\n",
       " 894: 'alpha macroglobulin',\n",
       " 895: 'Particles, Alpha',\n",
       " 896: 'Rhythm, Alpha',\n",
       " 897: 'alpha(S2)-Casein',\n",
       " 898: 'Globin, alpha-1',\n",
       " 899: 'alpha-1 Lipoprotein',\n",
       " 900: 'alpha Acetyldigitoxin',\n",
       " 901: 'alpha Acetyldigoxin',\n",
       " 902: 'alpha Actin',\n",
       " 903: 'Alpha Amanitin',\n",
       " 904: 'alpha Aminotoluene',\n",
       " 905: 'Alpha-Amylase',\n",
       " 906: 'alpha Bungarotoxin',\n",
       " 907: 'alpha-Casein',\n",
       " 908: 'alpha-CGRP',\n",
       " 909: 'alpha-Chlorhydrin',\n",
       " 910: 'Alpha Chymotrypsinogen',\n",
       " 911: 'alpha Collagen',\n",
       " 912: 'Crystallins, alpha',\n",
       " 913: 'Protease Inhibitors, alpha-Cysteine',\n",
       " 914: 'EFLORNITHINE',\n",
       " 915: 'alpha Elastin',\n",
       " 916: 'alpha Endosulfan',\n",
       " 917: 'alpha-D-galactosidase enzyme',\n",
       " 918: 'alpha Gliadin',\n",
       " 919: 'alpha globin',\n",
       " 920: 'alpha-glucosidase',\n",
       " 921: 'alpha-Hemocyanin',\n",
       " 922: 'alpha Isoactin',\n",
       " 923: 'alpha Kallikrein',\n",
       " 924: 'Lactalbumin alpha',\n",
       " 925: 'alpha Lactalbumin A',\n",
       " 926: 'alpha Lactalbumin B',\n",
       " 927: 'alpha Lactalbumin C',\n",
       " 928: 'Alpha-melanocyte stimulating hormone',\n",
       " 929: 'Alpha-Neurotoxin',\n",
       " 930: 'Alpha Peroxidase',\n",
       " 931: 'Toxin, alpha-Scorpion',\n",
       " 932: 'alpha-solanine',\n",
       " 933: 'alpha Spectrin',\n",
       " 934: 'Thalassemia-alpha',\n",
       " 935: 'alpha-Thrombin',\n",
       " 936: 'alpha Tropomyosin',\n",
       " 937: 'alpha Tubulin',\n",
       " 938: 'Alpha>2<-H foetoprotein',\n",
       " 939: 'Lipoproteins, alpha2-Very-Low-Density',\n",
       " 940: 'ALPHAPRODINE',\n",
       " 941: 'Algestone',\n",
       " 942: 'alphaviruses',\n",
       " 943: 'ALPRAZolam',\n",
       " 944: 'Alprénolol',\n",
       " 945: 'ALPROSTADIL',\n",
       " 946: 'ALSEROXYLON',\n",
       " 947: 'Alternaria (fungus)',\n",
       " 948: 'Alternate binaural loudness balance test',\n",
       " 949: 'Splicing, Alternative',\n",
       " 950: 'Medicine, Alternative',\n",
       " 951: 'Althesin',\n",
       " 952: 'altitudes',\n",
       " 953: 'Sickness, Altitude',\n",
       " 954: 'altruism',\n",
       " 955: 'Elements, Alu',\n",
       " 956: 'Alu-Like Repetitive Sequence',\n",
       " 957: 'Compounds, Alum',\n",
       " 958: 'Alumino 4 Aminosalicylic Acid',\n",
       " 959: 'Aluminum, NOS',\n",
       " 960: 'ammonium alum',\n",
       " 961: 'Aluminum and/or aluminum compound-containing product',\n",
       " 962: 'Aluminum aspirin',\n",
       " 963: 'ALUMINUM HYDROXIDE',\n",
       " 964: 'ALUMINUM HYDROXIDE GEL',\n",
       " 965: 'Nitrilotriacetate, Aluminum',\n",
       " 966: 'ALUMINUM OXIDE',\n",
       " 967: 'Silicates, Aluminum',\n",
       " 968: 'Bone Loss, Alveolar',\n",
       " 969: 'Process, Alveolar',\n",
       " 970: 'Ridge Augmentation, Alveolar',\n",
       " 971: 'alveolectomies',\n",
       " 972: 'Extrinsic alveolitis, allergic',\n",
       " 973: 'alveoloplasties',\n",
       " 974: \"Alzheimer's Diseases\",\n",
       " 975: 'Amanita, NOS',\n",
       " 976: 'Amanitin',\n",
       " 977: 'Amantidine',\n",
       " 978: 'Amantadine sulphate',\n",
       " 979: 'Amapari virus',\n",
       " 980: 'Amaranth Dye',\n",
       " 981: 'AMBENONIUM',\n",
       " 982: 'AMBENONIUM CHLORIDE',\n",
       " 983: 'ambidexterity',\n",
       " 984: 'Amblyopia NOS',\n",
       " 985: 'tobacco; amblyopia',\n",
       " 986: 'AMBROXOL',\n",
       " 987: 'ambulance',\n",
       " 988: 'Care, Ambulatory',\n",
       " 989: 'Facilities, Ambulatory Care',\n",
       " 990: 'Free-Standing Clinics',\n",
       " 991: 'Ambulatory Care Information Systems, Data Management',\n",
       " 992: 'Surgical Procedures, Ambulatory',\n",
       " 993: 'ambystoma',\n",
       " 994: 'mexicanums, Ambystoma',\n",
       " 995: 'Ambystomatidae',\n",
       " 996: 'Amdinocillin',\n",
       " 997: 'Amdinocillin pivoxil',\n",
       " 998: 'amoebiasis',\n",
       " 999: 'skin; amebiasis',\n",
       " ...}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main para debugear"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.10/dist-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/usr/local/lib/python3.10/dist-packages/spacy/language.py:2141: FutureWarning: Possible set union at position 6328\n",
      "  deserializers[\"tokenizer\"] = lambda p: self.tokenizer.from_disk(  # type: ignore[union-attr]\n",
      "/usr/local/lib/python3.10/dist-packages/sklearn/base.py:347: InconsistentVersionWarning: Trying to unpickle estimator TfidfTransformer from version 1.1.2 when using version 1.3.0. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.10/dist-packages/sklearn/base.py:347: InconsistentVersionWarning: Trying to unpickle estimator TfidfVectorizer from version 1.1.2 when using version 1.3.0. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import pickle as pkl\n",
    "from spacy_utils import *\n",
    "from entities import *\n",
    "from utils import *\n",
    "from graph import *\n",
    "from embeddings import *\n",
    "import networkx as nx\n",
    "from tqdm import tqdm\n",
    "\n",
    "desambiguation_entities_number = 6000000\n",
    "classes_number = 7000000\n",
    "NamesToInt = {}\n",
    "maximum_hopes = 3\n",
    "\n",
    "def expanded_graph(target, entity_tree, TUI_tree, ner, le):\n",
    "    if type(target) is not str and type(target) is not int:\n",
    "        for parent_entity in range(len(target)):\n",
    "            entity_tree.append([])\n",
    "            TUI_tree.append([])\n",
    "            expanded_graph(target[parent_entity], \n",
    "                           entity_tree[parent_entity], \n",
    "                           TUI_tree[parent_entity],\n",
    "                           ner, le)\n",
    "    else:\n",
    "        doc = ner.detected_entities(str(target))\n",
    "        list_linked_entities = ner.linked_entities(doc)\n",
    "        entities, TUIs = le.getNextEntitiesAndTUIs(target, list_linked_entities)\n",
    "        for entity in entities:\n",
    "            entity_tree.append(entity)\n",
    "        for TUI in TUIs:\n",
    "            TUI_tree.append(TUI)\n",
    "       \n",
    "    return entity_tree, TUI_tree\n",
    "\n",
    "def subgraph_generator(entities, graph, obtained_entities, TUIs, limit= 5):\n",
    "    obtained_entities = list(obtained_entities)\n",
    "    entity_list = []\n",
    "    triple_list = []\n",
    "    for entity in entities:\n",
    "        if entity in graph.nodes():\n",
    "            counter_in = 0\n",
    "            for triple in graph.in_edges(entity, data=True):\n",
    "                if triple[0] not in obtained_entities:\n",
    "                    target_TUIs = graph.nodes[triple[0]]['TUIs']\n",
    "                    for TUI in target_TUIs:\n",
    "                        if TUI in TUIs:\n",
    "                            if counter_in < limit:\n",
    "                                entity_list.append(triple[0])\n",
    "                                obtained_entities.append(triple[0])\n",
    "                                triple_list.append((triple[0], triple[2]['rel'], triple[1]))\n",
    "                                counter_in +=1\n",
    "                                break\n",
    "            counter_out = 0\n",
    "            for triple in graph.out_edges(entity, data=True):\n",
    "                if triple[1] not in obtained_entities:\n",
    "                    target_TUIs = graph.nodes[triple[1]]['TUIs']\n",
    "                    for TUI in target_TUIs:\n",
    "                        if TUI in TUIs:\n",
    "                            if counter_out < limit:\n",
    "                                entity_list.append(triple[1])\n",
    "                                obtained_entities.append(triple[1])\n",
    "                                triple_list.append((triple[0], triple[2]['rel'], triple[1]))\n",
    "                                counter_out +=1\n",
    "                                break\n",
    "    obtained_entities = list(set(obtained_entities))\n",
    "    entity_list = list(set(entity_list))\n",
    "    return entity_list, triple_list, obtained_entities\n",
    "\n",
    "\n",
    "with open('/usrvol/mappings/CUIsToInt.pickle', 'rb') as f:\n",
    "    CuisToInt = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/graph.pickle', 'rb') as f:\n",
    "    G = pkl.load(f)    \n",
    "\n",
    "with open('/usrvol/mappings/NamestoCUIFinal.pickle', 'rb') as f:\n",
    "    NamesToCUIs = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/IntToCUIs.pickle', 'rb') as f:\n",
    "    IntToCUIs = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/CUItoPreferredNamesFinal.pickle', 'rb') as f:\n",
    "    CUIToNames = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/IntToNames.pickle', 'rb') as f:\n",
    "    IntToNames = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/cuiToTUIs.pickle', 'rb') as f:\n",
    "    CUItoTUIs = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/TUIsToInt.pickle', 'rb') as f:\n",
    "    TUIToInt = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/RelationtoIdentifierTotal.pickle', 'rb') as f:\n",
    "    RelationToIdentifier = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/mapping_definitivo.pickle', 'rb') as f:\n",
    "    labelMapping = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/TUItoSemantic.pickle', 'rb') as f:\n",
    "    TUItoText = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/classes.pickle', 'rb') as f:\n",
    "    classes = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/classesToTextDict.pickle', 'rb') as f:\n",
    "    classesToText = pkl.load(f)\n",
    "\n",
    "mappers = (CuisToInt, IntToCUIs, NamesToCUIs, CUIToNames)\n",
    "\n",
    "# función recursiva que nos expande el grafo a partir de un target devolviendo un diccionario\n",
    "# con sus conexiones\n",
    "ner = Ner(spacy_model='en_core_sci_lg', \n",
    "          pipeline_linker='scispacy_linker', \n",
    "          resolve_abbreviations=True, \n",
    "          ontology=\"umls\", \n",
    "          threshold=\"0.85\", \n",
    "          max_entities_per_mention='100')\n",
    "\n",
    "le = Linked_Entities(resolve_abbreviations=True, ontology=\"umls\", threshold=0.85, max_entities_per_mention=100)\n",
    "\n",
    "text_path = '/usrvol/ISABIAL/Anotaciones_v2_ISABIAL/txt'\n",
    "entries = os.listdir(text_path)\n",
    "annotations_dict = {}\n",
    "vocab_mapper = {}\n",
    "iterator = 0\n",
    "counter_mapper = 0\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'expansion': 6000019,\n",
       " 'huntingtin gene': 6000020,\n",
       " 'Phosphorylation': 6000021,\n",
       " 'translation initiation factor': 6000022,\n",
       " 'alteration': 6000023,\n",
       " 'protein synthesis': 6000024,\n",
       " 'age': 6000025,\n",
       " 'factor': 6000026,\n",
       " 'adult': 6000027,\n",
       " 'factors': 6000028,\n",
       " 'maternal': 6000029,\n",
       " 'diagnosis': 6000030,\n",
       " 'care': 6000031,\n",
       " 'primary': 6000032,\n",
       " 'review': 6000033,\n",
       " 'genetics': 6000034,\n",
       " 'pathology': 6000035,\n",
       " 'imaginal disc-derived wing expansion': 6000036,\n",
       " 'genetic testing': 6000037}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NamesToInt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(NamesToInt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'imaginal disc-derived wing expansion'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IntToNames[1023783]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'imaginal disc-derived wing expansion' in NamesToInt.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6000017"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NamesToInt['imaginal disc-derived wing expansion']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "NamesToInt = {}\n",
    "annotations_dict = {}\n",
    "vocab_mapper = {}\n",
    "iterator = 0\n",
    "counter_mapper = 0\n",
    "for class_ in classes:\n",
    "    IntToNames[classes_number] = classesToText[class_]\n",
    "    NamesToInt[class_] = classes_number\n",
    "    vocab_mapper[classes_number] = counter_mapper\n",
    "    classes_number += 1\n",
    "    counter_mapper += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 1/1000 [00:00<02:40,  6.24it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 0\n",
      "file number: 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 3/1000 [00:00<02:15,  7.34it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 2\n",
      "file number: 3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 5/1000 [00:00<02:44,  6.04it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 4\n",
      "file number: 5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|          | 7/1000 [00:01<02:25,  6.83it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 6\n",
      "file number: 7\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|          | 8/1000 [00:01<02:15,  7.31it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|          | 10/1000 [00:01<02:35,  6.35it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 9\n",
      "file number: 10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|          | 11/1000 [00:01<02:45,  5.99it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 11\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|▏         | 13/1000 [00:02<03:00,  5.46it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 12\n",
      "file number: 13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  2%|▏         | 15/1000 [00:02<02:54,  5.64it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 14\n",
      "file number: 15\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  2%|▏         | 17/1000 [00:02<02:43,  6.00it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 16\n",
      "file number: 17\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  2%|▏         | 19/1000 [00:03<02:29,  6.57it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 18\n",
      "file number: 19\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  2%|▏         | 21/1000 [00:03<02:32,  6.42it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 20\n",
      "file number: 21\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  2%|▏         | 23/1000 [00:03<02:27,  6.64it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 22\n",
      "file number: 23\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  2%|▎         | 25/1000 [00:04<02:47,  5.81it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 24\n",
      "file number: 25\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  3%|▎         | 27/1000 [00:04<02:40,  6.06it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 26\n",
      "file number: 27\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  3%|▎         | 29/1000 [00:04<02:57,  5.48it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 28\n",
      "file number: 29\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  3%|▎         | 30/1000 [00:04<02:54,  5.57it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 30\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  3%|▎         | 32/1000 [00:05<03:25,  4.72it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 31\n",
      "file number: 32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  4%|▎         | 35/1000 [00:05<02:20,  6.88it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 33\n",
      "file number: 34\n",
      "file number: 35\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  4%|▎         | 36/1000 [00:05<02:27,  6.55it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 36\n",
      "file number: 37\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  4%|▍         | 39/1000 [00:06<02:28,  6.46it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 38\n",
      "file number: 39\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  4%|▍         | 41/1000 [00:06<02:13,  7.17it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 40\n",
      "file number: 41\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  4%|▍         | 43/1000 [00:06<02:01,  7.90it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 42\n",
      "file number: 43\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  4%|▍         | 45/1000 [00:07<02:29,  6.38it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 44\n",
      "file number: 45\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  5%|▍         | 47/1000 [00:07<02:30,  6.35it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 46\n",
      "file number: 47\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  5%|▍         | 49/1000 [00:07<02:28,  6.39it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 48\n",
      "file number: 49\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  5%|▌         | 51/1000 [00:08<02:26,  6.46it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 50\n",
      "file number: 51\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  5%|▌         | 52/1000 [00:08<02:45,  5.74it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 52\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  5%|▌         | 54/1000 [00:08<02:57,  5.32it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 53\n",
      "file number: 54\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  6%|▌         | 56/1000 [00:09<02:53,  5.45it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 55\n",
      "file number: 56\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  6%|▌         | 58/1000 [00:09<02:23,  6.54it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 57\n",
      "file number: 58\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  6%|▌         | 60/1000 [00:09<02:26,  6.43it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 59\n",
      "file number: 60\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  6%|▌         | 61/1000 [00:09<02:18,  6.77it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 61\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  6%|▌         | 62/1000 [00:10<03:59,  3.92it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 62\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  6%|▋         | 63/1000 [00:10<03:45,  4.16it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 63\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  6%|▋         | 64/1000 [00:10<04:15,  3.66it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  6%|▋         | 65/1000 [00:11<04:13,  3.69it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 65\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  7%|▋         | 67/1000 [00:11<03:19,  4.68it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file number: 66\n",
      "file number: 67\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  7%|▋         | 68/1000 [00:11<03:17,  4.72it/s]"
     ]
    }
   ],
   "source": [
    "for file_path in tqdm(range(1000)):\n",
    "    iterator += 1\n",
    "    print(f\"file number: {file_path}\" )\n",
    "    if f\"text{file_path}.txt\" in entries:\n",
    "        file = f\"text{file_path}.txt\"\n",
    "        with open(f\"/usrvol/ISABIAL/Anotaciones_v2_ISABIAL/txt/{file}\", 'r') as f:\n",
    "            lines = f.readlines()\n",
    "        text = lines[0]\n",
    "        candidates_dict = {}\n",
    "        doc = ner.detected_entities(text)\n",
    "        list_linked_entities = ner.linked_entities(doc)\n",
    "        candidates, candidates_TUIs = le.candidatesChecker(list_linked_entities)\n",
    "        \n",
    "        # Con esto evitamos repetir candidatos si aparecen varias veces en el texto\n",
    "        new_candidates = []\n",
    "        for candidate in candidates:\n",
    "            if candidate not in new_candidates:\n",
    "                new_candidates.append(candidate)\n",
    "        candidates = new_candidates\n",
    "        \n",
    "        \n",
    "        entity_list, TUI_list = le.getEntitiesAndTUIs(list_linked_entities)\n",
    "        final_graphs = {}\n",
    "        labels = {}\n",
    "\n",
    "        # Creamos una lista de Nombres de entidades a desambiguar mapeandolas a un entero\n",
    "        for candidate in range(len(candidates)):     \n",
    "            if not candidates[candidate] in NamesToInt.keys():\n",
    "                IntToNames[desambiguation_entities_number] = candidates[candidate]\n",
    "                NamesToInt[candidates[candidate]] = desambiguation_entities_number\n",
    "                vocab_mapper[desambiguation_entities_number] = counter_mapper\n",
    "                desambiguation_entities_number += 1\n",
    "                counter_mapper +=1\n",
    "            \n",
    "            else:\n",
    "                # Si se cumple esta condición quiere decir que el término aparece tanto como candidato como en el vocabulario original\n",
    "                if NamesToInt[candidates[candidate]] not in vocab_mapper.keys():\n",
    "                    vocab_mapper[NamesToInt[candidates[candidate]]] = counter_mapper\n",
    "                    counter_mapper +=1\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        # iteramos sobre cada candidato teniendo ya el vocabulario de los candidatos   \n",
    "        for candidate in range(len(candidates)):\n",
    "            graph_entities = {'first_entities': [], 'last_entities': [], 'TUIs': []}\n",
    "            obtained_entities = []\n",
    "            TUIs = candidates_TUIs[candidate]\n",
    "            forward_triples_list = []\n",
    "            backward_triples_list = []\n",
    "            connected_entities, _ = expanded_graph(candidates[candidate], [], [], ner, le)\n",
    "\n",
    "            # Nos quedamos solo con las entidades que estén en el vocabulario según la base de datos\n",
    "            newconnected_entities = []\n",
    "            for entity in range(len(connected_entities)):\n",
    "                    if connected_entities[entity] in NamesToCUIs.keys():\n",
    "                        newconnected_entities.append(CuisToInt[NamesToCUIs[connected_entities[entity].lower()]])\n",
    "                        \n",
    "                        # Añadimos estas entidades al vocabulario\n",
    "                        if not connected_entities[entity] in NamesToInt.keys():\n",
    "                            IntToNames[desambiguation_entities_number] = connected_entities[entity]\n",
    "                            NamesToInt[connected_entities[entity]] = desambiguation_entities_number\n",
    "                            vocab_mapper[desambiguation_entities_number] = counter_mapper\n",
    "                            desambiguation_entities_number += 1\n",
    "                            counter_mapper +=1\n",
    "            \n",
    "            connected_entities = newconnected_entities\n",
    "            graph_entities['first_entities'] = connected_entities\n",
    "  \n",
    "            # formamos el grafo desde las entidades iniciales\n",
    "            for hop in range(maximum_hopes):\n",
    "                connected_entities, triples, obtained_entities = subgraph_generator(connected_entities, G, obtained_entities, TUIs)\n",
    "                forward_triples_list.extend(triples)\n",
    "                forward_triples_list = list(set(forward_triples_list))\n",
    "            Rconnected_entities, RTUIs = le.otherRelatedEntities(candidates[candidate], \n",
    "                                                        candidates, \n",
    "                                                        candidates_TUIs, \n",
    "                                                        entity_list, TUI_list)\n",
    "            \n",
    "            # Nos quedamos solo con las entidades que estén en el vocabulario según la base de datos\n",
    "            newRconnected_entities = []\n",
    "            for entity in range(len(Rconnected_entities)):\n",
    "                if Rconnected_entities[entity] in NamesToCUIs.keys():\n",
    "                    if not Rconnected_entities[entity] in NamesToInt.keys():\n",
    "                        IntToNames[desambiguation_entities_number] = Rconnected_entities[entity]\n",
    "                        NamesToInt[Rconnected_entities[entity]] = desambiguation_entities_number\n",
    "                        vocab_mapper[desambiguation_entities_number] = counter_mapper\n",
    "                        desambiguation_entities_number += 1\n",
    "                        counter_mapper +=1\n",
    "\n",
    "                    if NamesToCUIs[Rconnected_entities[entity].lower()] not in CuisToInt.keys():\n",
    "                        CuisToInt[NamesToCUIs[Rconnected_entities[entity].lower()]] = counter_mapper\n",
    "                    newRconnected_entities.append(CuisToInt[NamesToCUIs[Rconnected_entities[entity].lower()]])\n",
    "            Rconnected_entities = newRconnected_entities\n",
    "            graph_entities['last_entities'] = Rconnected_entities\n",
    "            graph_entities['TUIs'] = RTUIs\n",
    "            statements_graphs = []\n",
    "           \n",
    "            # añadimos los TUIs a labels para usarlos como etiquetas\n",
    "            for TUI in TUIs:\n",
    "                if not candidates[candidate] in labels.keys():\n",
    "                    labels[candidates[candidate]] = [labelMapping[TUI]]\n",
    "                else:\n",
    "                    labels[candidates[candidate]].append(labelMapping[TUI])\n",
    "\n",
    "            # Formamos todas las variantes de grafo desde las entidades finales\n",
    "            for class_ in classes:\n",
    "                if class_ in labels.keys():\n",
    "                    graph_last_entities = []\n",
    "\n",
    "                    # Si la entidad final de la lista de entidades contiene entre sus TUIs el TUI objetivo...\n",
    "                    for entity in range(len(graph_entities['last_entities'])):\n",
    "                        if TUI in graph_entities['TUIs'][entity]:\n",
    "                            # Se añade a graph_last_entities para formar este grafo en particular\n",
    "                            graph_last_entities.append(graph_entities['last_entities'][entity])\n",
    "                    final_entities = graph_last_entities\n",
    "                    # se obtienen las tripletas inversas para este TUI en particular\n",
    "                    for hop in range(maximum_hopes):      \n",
    "                        graph_last_entities, triples, obtained_entities = subgraph_generator(graph_last_entities, G, obtained_entities, TUIs)\n",
    "                        backward_triples_list.extend(triples)\n",
    "                        backward_triples_list = list(set(backward_triples_list))\n",
    "                    triples_list = forward_triples_list + backward_triples_list\n",
    "                    graph = nx.MultiDiGraph()\n",
    "                    target = candidates[candidate]\n",
    "                    initial_entities = graph_entities['first_entities']\n",
    "                    final_entities = graph_entities['last_entities']\n",
    "                    attrs = set()\n",
    "                    for triple in triples_list:\n",
    "                        head = triple[0]\n",
    "                        # Añadimos todas las entidades y relaciones de las tripletas de cada grafo al vocabulario si no existen aún\n",
    "                        if not IntToNames[head] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[head]] = head\n",
    "                            vocab_mapper[head] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "                        elif not head in vocab_mapper.keys():\n",
    "                            head = NamesToInt[IntToNames[head]]\n",
    "\n",
    "                        tail = triple[2]\n",
    "                        # Idem\n",
    "                        if not IntToNames[tail] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[tail]] = tail\n",
    "                            vocab_mapper[tail] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "                        elif not tail in vocab_mapper.keys():\n",
    "                            tail = NamesToInt[IntToNames[tail]]\n",
    "                        \n",
    "                        graph.add_node(vocab_mapper[head], TUIs='None', node_type=2, node_score=0)\n",
    "                        graph.add_node(vocab_mapper[tail], TUIs='None', node_type=2, node_score=0)\n",
    "                        rel = triple[1]\n",
    "                        weight = 1.\n",
    "                        graph.add_edge(vocab_mapper[head], vocab_mapper[tail], rel=rel, weight=weight)\n",
    "                        attrs.add((vocab_mapper[head], vocab_mapper[tail], rel))\n",
    "                    # añadimos el nodo inicial y sus conexiones\n",
    "                    graph.add_node(vocab_mapper[NamesToInt[target]], TUIs=TUIs, node_type=0, node_score=0)\n",
    "                    for entity in initial_entities:\n",
    "                        if not IntToNames[entity] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[entity]] = entity\n",
    "                            vocab_mapper[entity] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "                        elif not entity in vocab_mapper.keys():\n",
    "                            entity = NamesToInt[IntToNames[entity]]\n",
    "                        if not entity in graph.nodes:\n",
    "                            try:\n",
    "                                T = CUItoTUIs[IntToCUIs[entity]]\n",
    "                            except:\n",
    "                                T = 'NOT FOUND'\n",
    "                        graph.add_node(vocab_mapper[entity], TUIs=[T], node_type=2, node_score=0)\n",
    "                        graph.add_edge(vocab_mapper[NamesToInt[target]], vocab_mapper[entity], rel=RelationToIdentifier['meaning_of_concept'], weight=1.)\n",
    "                    # añadimos el nodo final y sus conexiones\n",
    "                    graph.add_node(vocab_mapper[NamesToInt[class_]], TUIs=['None'], node_type=1, node_score=0)\n",
    "                    for entity in final_entities:\n",
    "                        if not IntToNames[entity] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[entity]] = entity\n",
    "                            vocab_mapper[entity] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "                        elif not entity in vocab_mapper.keys():\n",
    "                            entity = NamesToInt[IntToNames[entity]]\n",
    "                        if not entity in graph.nodes:\n",
    "                            try:\n",
    "                                T = CUItoTUIs[IntToCUIs[entity]]\n",
    "                            except:\n",
    "                                T = 'NOT FOUND'\n",
    "                        graph.add_node(vocab_mapper[entity], TUIs=['None'], node_type=2, node_score=0)\n",
    "                        graph.add_edge(vocab_mapper[entity], vocab_mapper[NamesToInt[class_]], rel=RelationToIdentifier['belongs_to'], weight=1.)\n",
    "                    statements_graphs.append(graph)\n",
    "                    backward_triples_list = []\n",
    "                \n",
    "                else:\n",
    "                    triples_list = forward_triples_list\n",
    "                    graph = nx.MultiDiGraph()\n",
    "                    target = candidates[candidate]\n",
    "                    initial_entities = graph_entities['first_entities']\n",
    "                    attrs = set()\n",
    "                    for triple in triples_list:\n",
    "                        \n",
    "                        \n",
    "                        head = triple[0]\n",
    "                        # Añadimos todas las entidades y relaciones de las tripletas de cada grafo al vocabulario si no existen aún\n",
    "                        if not IntToNames[head] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[head]] = head\n",
    "                            vocab_mapper[head] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "                        # Esto quiere decir que el término aparece tanto como candidato como en el vocabulario original, teniendo dos números diferentes que vamos a mapear a uno solo\n",
    "                        elif not head in vocab_mapper.keys():\n",
    "                            head = NamesToInt[IntToNames[head]]\n",
    "                        \n",
    "                        \n",
    "                        tail = triple[2]\n",
    "                        # Idem\n",
    "                        if not IntToNames[tail] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[tail]] = tail\n",
    "                            vocab_mapper[tail] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "                        # Esto quiere decir que el término aparece tanto como candidato como en el vocabulario original, teniendo dos números diferentes que vamos a mapear a uno solo\n",
    "                        elif not tail in vocab_mapper.keys():\n",
    "                            tail = NamesToInt[IntToNames[tail]]\n",
    "\n",
    "\n",
    "                        graph.add_node(vocab_mapper[head], TUIs='none', node_type=2, node_score=0)\n",
    "                        graph.add_node(vocab_mapper[tail], TUIs='none', node_type=2, node_score=0)\n",
    "                        rel = triple[1]\n",
    "                        weight = 1.\n",
    "                        graph.add_edge(vocab_mapper[head], vocab_mapper[tail], rel=rel, weight=weight)\n",
    "                        attrs.add((vocab_mapper[head], vocab_mapper[tail], rel))\n",
    "                    # añadimos el nodo inicial y sus conexiones\n",
    "                    graph.add_node(vocab_mapper[NamesToInt[target]], TUIs=TUIs, node_type=0, node_score=0)\n",
    "                    for entity in initial_entities:\n",
    "                        if not IntToNames[entity] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[entity]] = entity\n",
    "                            vocab_mapper[entity] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "\n",
    "                        # Esto quiere decir que el término aparece tanto como candidato como en el vocabulario original, teniendo dos números diferentes que vamos a mapear a uno solo\n",
    "                        elif not entity in vocab_mapper.keys():\n",
    "                            entity = NamesToInt[IntToNames[entity]]\n",
    "                        if not entity in graph.nodes:\n",
    "                            try:\n",
    "                                T = CUItoTUIs[IntToCUIs[entity]]\n",
    "                            except:\n",
    "                                T = 'NOT FOUND'     \n",
    "                        graph.add_node(vocab_mapper[entity], TUIs=[T], node_type=2, node_score=0)\n",
    "                        graph.add_edge(vocab_mapper[NamesToInt[target]], vocab_mapper[entity], rel=RelationToIdentifier['meaning_of_concept'], weight=1.)\n",
    "                    # añadimos el nodo final y sin conexiones porque no es su clase\n",
    "                    graph.add_node(vocab_mapper[NamesToInt[class_]], TUIs=['None'], node_type=1, node_score=0)\n",
    "                    statements_graphs.append(graph)           \n",
    "            final_graphs[candidates[candidate]] = statements_graphs\n",
    "\n",
    "\n",
    "        # Debemos eliminar todas las opciones que aparezcan con un NONE\n",
    "        candidates_list = []\n",
    "        for key in labels.keys():\n",
    "            new_labels = []\n",
    "            new_graphs = []\n",
    "            for i in range(len(labels[key])):\n",
    "                if labels[key][i] != 'NONE':\n",
    "                    new_labels.append(labels[key][i])\n",
    "            new_labels = list(set(new_labels))\n",
    "            if len(new_labels) < 1:\n",
    "                candidates_list.append(key)\n",
    "            labels[key] = new_labels\n",
    "            final_graphs[key] = new_graphs\n",
    "\n",
    "        for key in candidates_list:\n",
    "            del labels[key]\n",
    "\n",
    "        with open(f\"/usrvol/data/categorical/labels{iterator}.pickle\", 'wb') as f:\n",
    "            pkl.dump(labels, f)\n",
    "            \n",
    "        with open(f\"/usrvol/data/categorical/final_graphs{iterator}.pickle\", 'wb') as f:\n",
    "            pkl.dump(final_graphs, f)\n",
    "\n",
    "\n",
    "# En lugar de usar vocab_dict mapeamos con vocab_mapper usando intToNames\n",
    "reverse_vocab_mapper = {v: k for k, v in vocab_mapper.items()}\n",
    "with open(\"/usrvol/mappings/categorical/vocab.txt\", 'w') as f:\n",
    "    for i in range(len(reverse_vocab_mapper)):\n",
    "        value = IntToNames[reverse_vocab_mapper[i]]\n",
    "        print(value, file=f)\n",
    "\n",
    "with open('/usrvol/mappings/categorical/vocab_mapper.pickle', 'wb') as f:\n",
    "            pkl.dump(vocab_mapper, f)\n",
    "print(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "547367",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m/usrvol/GitHub2/debug_categorical.ipynb Cell 23\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f72656c617865645f63686174656c6574227d/usrvol/GitHub2/debug_categorical.ipynb#X34sdnNjb2RlLXJlbW90ZQ%3D%3D?line=0'>1</a>\u001b[0m vocab_mapper[\u001b[39m547367\u001b[39;49m]\n",
      "\u001b[0;31mKeyError\u001b[0m: 547367"
     ]
    }
   ],
   "source": [
    "vocab_mapper[547367]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'landscape'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IntToNames[547367]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6000057"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NamesToInt['actin phosphorylation']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6003346"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NamesToInt[IntToNames[547367]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotation dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle as pkl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/usrvol/data/categorical/final_labels.pickle', 'rb') as f:\n",
    "    final_labels = pkl.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: {'huntingtin gene': ['GeneOrGenome'],\n",
       "  'Phosphorylation': ['CellFunction'],\n",
       "  'translation initiation factor': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'protein synthesis': ['CellFunction'],\n",
       "  'adult': ['PopulationGroup'],\n",
       "  'maternal': ['PopulationGroup'],\n",
       "  'diagnosis': ['HealthCareActivity'],\n",
       "  'genetics': ['BiomedicalOccupationOrDiscipline'],\n",
       "  'pathology': ['PathologicFunction', 'BiomedicalOccupationOrDiscipline']},\n",
       " 2: {'muscle stiffness': ['SignOrSymptom'],\n",
       "  'symptoms': ['SignOrSymptom'],\n",
       "  'dopamine': ['BiologicallyActiveSubstance']},\n",
       " 3: {'Mitochondria': ['CellComponent'],\n",
       "  'health': ['HealthCareActivity'],\n",
       "  'metabolic process': ['CellFunction'],\n",
       "  'mitochondria': ['CellComponent']},\n",
       " 5: {'symptom': ['SignOrSymptom'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'pathophysiology': ['PathologicFunction']},\n",
       " 7: {\"Alzheimer's\": ['DiseaseOrSyndrome'],\n",
       "  'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'quality of life': ['Finding'],\n",
       "  'treatment': ['HealthCareActivity']},\n",
       " 10: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'DNA damage repair': ['CellFunction'],\n",
       "  'age': ['OrganismAttribute'],\n",
       "  'DNA repair genes': ['GeneOrGenome']},\n",
       " 11: {'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'clinical studies': ['ResearchActivity'],\n",
       "  'approval': ['ResearchActivity'],\n",
       "  'guidance': ['ClinicalAttribute']},\n",
       " 12: {'visual dysfunction': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome']},\n",
       " 13: {'death': ['BiologicFunction'],\n",
       "  'mental functioning': ['BiologicFunction'],\n",
       "  'peripheral nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'prevention': ['HealthCareActivity'],\n",
       "  'public health': ['HealthCareActivity', 'BiomedicalOccupationOrDiscipline'],\n",
       "  'Melatonin': ['Chemical',\n",
       "   'PharmacologicSubstance',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'melatonin': ['Chemical',\n",
       "   'PharmacologicSubstance',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'signaling pathways': ['CellFunction']},\n",
       " 14: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'aggregation': ['CellFunction'],\n",
       "  'treatments': ['HealthCareActivity'],\n",
       "  'review': ['ResearchActivity'],\n",
       "  'proteasome': ['AminoAcidPeptideOrProtein']},\n",
       " 15: {'communication': ['HealthCareActivity', 'CellFunction'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'discussions': ['HealthCareActivity'],\n",
       "  'documentation': ['HealthCareActivity']},\n",
       " 16: {'long-term': ['TemporalConcept'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'participants': ['PopulationGroup'],\n",
       "  'quality-of-life': ['Finding']},\n",
       " 17: {'treatments': ['HealthCareActivity'],\n",
       "  'risk': ['Finding'],\n",
       "  'age': ['OrganismAttribute'],\n",
       "  'concomitant': ['TemporalConcept'],\n",
       "  'population': ['PopulationGroup'],\n",
       "  'individuals': ['PopulationGroup'],\n",
       "  'improved': ['Finding'],\n",
       "  'Inflammation': ['PathologicFunction'],\n",
       "  'nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'CNS': ['BodyPartOrganOrOrganComponent'],\n",
       "  'resident': ['PopulationGroup'],\n",
       "  'microenvironment': ['BodyPartOrganOrOrganComponent',\n",
       "   'NaturalPhenomenonOrProcess'],\n",
       "  'behavior': ['BiologicFunction', 'IndividualBehavior'],\n",
       "  'toxin': ['InjuryOrPoisoning', 'BiologicallyActiveSubstance'],\n",
       "  'Amyotrophic Lateral Sclerosis': ['DiseaseOrSyndrome']},\n",
       " 18: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'drug development': ['ResearchActivity'],\n",
       "  'pharmaceutical industry': ['HealthCareRelatedOrganization']},\n",
       " 25: {'proteins': ['AminoAcidPeptideOrProtein'],\n",
       "  'aged': ['TemporalConcept'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'treatment': ['HealthCareActivity']},\n",
       " 27: {'population': ['PopulationGroup'],\n",
       "  'sleep disorders': ['DiseaseOrSyndrome'],\n",
       "  'quality of life': ['Finding'],\n",
       "  'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'adverse effects': ['PathologicFunction'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'recognition': ['BiologicFunction'],\n",
       "  'exercise': ['HealthCareActivity'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome']},\n",
       " 28: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'tau': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'alpha-synuclein': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'huntingtin': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'prion proteins': ['AminoAcidPeptideOrProtein'],\n",
       "  'angiogenesis': ['BiologicFunction'],\n",
       "  'biomarkers': ['ClinicalAttribute']},\n",
       " 29: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'autosomal dominant': ['CellFunction'],\n",
       "  'Huntingtin gene': ['GeneOrGenome'],\n",
       "  'symptoms': ['SignOrSymptom'],\n",
       "  'suicide': ['Finding'],\n",
       "  'severity': ['Finding'],\n",
       "  'illness': ['DiseaseOrSyndrome'],\n",
       "  'data': ['ResearchActivity']},\n",
       " 30: {'self-regulation': ['BiologicFunction', 'IndividualBehavior']},\n",
       " 34: {'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'dysfunction': ['PathologicFunction'],\n",
       "  'death': ['BiologicFunction'],\n",
       "  'pathway': ['CellFunction']},\n",
       " 37: {'Peroxiredoxins': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'mammalian': ['Eukaryote'],\n",
       "  'studies': ['ResearchActivity'],\n",
       "  'enzymes': ['BiologicallyActiveSubstance'],\n",
       "  'knowledge': ['Finding'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'Lewy bodies': ['CellComponent'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome']},\n",
       " 41: {'cellular homeostasis': ['CellFunction'],\n",
       "  'molecular chaperones': ['AminoAcidPeptideOrProtein'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'alpha-synuclein': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'relationship': ['Finding'],\n",
       "  'treatment': ['HealthCareActivity']},\n",
       " 46: {'humans': ['Eukaryote'],\n",
       "  \"Alzheimer's Disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'Amyotrophic Lateral Sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'disease progression': ['PathologicFunction'],\n",
       "  'knowledge': ['Finding']},\n",
       " 48: {'neuronal degeneration': ['CellOrMolecularDysfunction'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'abnormal': ['Finding'],\n",
       "  'connexins': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'glia': ['Cell'],\n",
       "  'ATP': ['PharmacologicSubstance']},\n",
       " 62: {'research': ['ResearchActivity'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'progression': ['PathologicFunction', 'TemporalConcept']},\n",
       " 67: {'DNA methylation': ['CellFunction'],\n",
       "  'methylation': ['CellFunction'],\n",
       "  'demethylation': ['CellFunction'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome']},\n",
       " 68: {'symptomatic': ['SignOrSymptom']},\n",
       " 77: {'huntingtin gene': ['GeneOrGenome'],\n",
       "  'expressed': ['CellFunction'],\n",
       "  'gastrointestinal system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'pathway': ['CellFunction'],\n",
       "  'gut': ['BodyPartOrganOrOrganComponent']},\n",
       " 78: {'toxic': ['InjuryOrPoisoning'],\n",
       "  'V(D)J recombination': ['CellFunction'],\n",
       "  'meiotic recombination': ['CellFunction'],\n",
       "  'DNA replication': ['CellFunction'],\n",
       "  'transcription': ['CellFunction'],\n",
       "  'human': ['Eukaryote'],\n",
       "  'dysfunction': ['PathologicFunction'],\n",
       "  'repair': ['HealthCareActivity', 'BiologicFunction'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome']},\n",
       " 79: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'p53': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'transcription factor': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'occurrence': ['TemporalConcept'],\n",
       "  'research': ['ResearchActivity']},\n",
       " 81: {'therapeutic trial': ['ResearchActivity'],\n",
       "  'patient': ['PatientOrDisabledGroup'],\n",
       "  'guidance': ['ClinicalAttribute'],\n",
       "  'clinical trials': ['ResearchActivity']},\n",
       " 82: {'morbidity': ['Finding'],\n",
       "  'symptoms': ['SignOrSymptom'],\n",
       "  'pathophysiology': ['PathologicFunction'],\n",
       "  \"Alzheimer's\": ['DiseaseOrSyndrome'],\n",
       "  'CNS': ['BodyPartOrganOrOrganComponent'],\n",
       "  'endocannabinoids': ['Chemical', 'BiologicallyActiveSubstance'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'MS': ['DiseaseOrSyndrome']},\n",
       " 85: {'research': ['ResearchActivity'],\n",
       "  'prion protein': ['AminoAcidPeptideOrProtein'],\n",
       "  'studies': ['ResearchActivity'],\n",
       "  \"Alzheimer's\": ['DiseaseOrSyndrome'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent']},\n",
       " 87: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'encoding': ['BiologicFunction'],\n",
       "  'huntingtin protein': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'motor dysfunction': ['DiseaseOrSyndrome']},\n",
       " 91: {\"Alzheimer's\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'CNS': ['BodyPartOrganOrOrganComponent'],\n",
       "  'initiation': ['TemporalConcept'],\n",
       "  'development': ['CellFunction', 'BiologicFunction'],\n",
       "  'Mitochondria': ['CellComponent'],\n",
       "  'cellular metabolism regulation': ['CellFunction'],\n",
       "  'intracellular': ['CellComponent'],\n",
       "  'immune responses': ['BiologicFunction'],\n",
       "  'cell fate': ['CellFunction']},\n",
       " 98: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'Prodynorphin': ['AminoAcidPeptideOrProtein'],\n",
       "  'Postmortem': ['TemporalConcept']},\n",
       " 99: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'disability': ['Finding'],\n",
       "  'patient': ['PatientOrDisabledGroup'],\n",
       "  'independence': ['BiologicFunction']},\n",
       " 100: {'exon': ['NucleicAcidNucleosideOrNucleotide'],\n",
       "  'recovery': ['BiologicFunction'],\n",
       "  'Exendin 4': ['AminoAcidPeptideOrProtein'],\n",
       "  'CaM': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'BiP': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'mutated protein': ['AminoAcidPeptideOrProtein'],\n",
       "  'peptides': ['AminoAcidPeptideOrProtein']},\n",
       " 101: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'huntingtin gene': ['GeneOrGenome'],\n",
       "  'studies': ['ResearchActivity'],\n",
       "  'age': ['OrganismAttribute']},\n",
       " 102: {'Parkinson?s disease': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'Alzheimer?s disease': ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'Amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'MS': ['DiseaseOrSyndrome'],\n",
       "  'year': ['TemporalConcept'],\n",
       "  'risk factor': ['Finding'],\n",
       "  'flavonols': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'treatment': ['HealthCareActivity']},\n",
       " 103: {'manifestations': ['SignOrSymptom'],\n",
       "  'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'neutropenia': ['DiseaseOrSyndrome']},\n",
       " 104: {'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'knowledge': ['Finding'],\n",
       "  'identification': ['BiologicFunction'],\n",
       "  'matrix metalloproteinases': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'MMP': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'ADAM': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'ADAMTS': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'metalloproteinases': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance']},\n",
       " 105: {'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'encoding': ['BiologicFunction'],\n",
       "  'susceptible': ['OrganismAttribute']},\n",
       " 106: {'Berberine': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'berberine': ['Chemical', 'PharmacologicSubstance']},\n",
       " 107: {'treatment': ['HealthCareActivity'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'encoding': ['BiologicFunction'],\n",
       "  'zinc-finger nucleases': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'transcription activator-like effector nucleases': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'TALENs': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'gene editing': ['ResearchActivity'],\n",
       "  'cell types': ['Cell'],\n",
       "  'human': ['Eukaryote'],\n",
       "  'Cas9': ['BiologicallyActiveSubstance']},\n",
       " 108: {'human': ['Eukaryote'],\n",
       "  'zebrafish': ['Eukaryote'],\n",
       "  'Danio rerio': ['Eukaryote'],\n",
       "  'Oryzias latipes': ['Eukaryote'],\n",
       "  'gene knockout': ['ResearchActivity'],\n",
       "  'fish': ['Eukaryote']},\n",
       " 109: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'nestin': ['GeneOrGenome'],\n",
       "  'Differentiation': ['CellFunction'],\n",
       "  'immunocytochemistry': ['ResearchActivity'],\n",
       "  'treatment': ['ResearchActivity']},\n",
       " 110: {'morbidity': ['Finding'],\n",
       "  'treatments': ['HealthCareActivity'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'biomarkers': ['ClinicalAttribute'],\n",
       "  'research': ['ResearchActivity'],\n",
       "  'lncRNA': ['NucleicAcidNucleosideOrNucleotide']},\n",
       " 111: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'degeneration': ['PathologicFunction'],\n",
       "  'nervous system': ['BodyPartOrganOrOrganComponent']},\n",
       " 112: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'mutation': ['CellOrMolecularDysfunction'],\n",
       "  'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'adaptation': ['BiologicFunction']},\n",
       " 113: {'HD': ['DiseaseOrSyndrome'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'clearance': ['NaturalPhenomenonOrProcess'],\n",
       "  'mitochondria': ['CellComponent'],\n",
       "  'review': ['HealthCareActivity']},\n",
       " 114: {'microRNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'small non-coding RNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'microRNA': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ataxia': ['DiseaseOrSyndrome'],\n",
       "  'glaucoma': ['DiseaseOrSyndrome']},\n",
       " 115: {'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'age': ['OrganismAttribute'],\n",
       "  'disease onset': ['TemporalConcept'],\n",
       "  'protein aggregation': ['CellFunction'],\n",
       "  'longevity': ['TemporalConcept']},\n",
       " 116: {'reactive oxygen species': ['Chemical', 'BiologicallyActiveSubstance'],\n",
       "  'chronic': ['TemporalConcept'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'brain pathologies': ['PathologicFunction'],\n",
       "  'polyphenols': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'carotenes': ['Chemical', 'BiologicallyActiveSubstance'],\n",
       "  'vitamins': ['BiologicallyActiveSubstance'],\n",
       "  'hormones': ['BiologicallyActiveSubstance'],\n",
       "  'melatonin': ['Chemical',\n",
       "   'PharmacologicSubstance',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome']},\n",
       " 117: {'cognitive impairment': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome']},\n",
       " 118: {'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'patient': ['PatientOrDisabledGroup'],\n",
       "  'testing': ['HealthCareActivity']},\n",
       " 119: {'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'Studies': ['ResearchActivity'],\n",
       "  'recall': ['BiologicFunction'],\n",
       "  'recognition': ['BiologicFunction']},\n",
       " 120: {'research': ['ResearchActivity'],\n",
       "  'epidemiology': ['BiomedicalOccupationOrDiscipline'],\n",
       "  'frequency': ['TemporalConcept'],\n",
       "  'race': ['PopulationGroup'],\n",
       "  'ethnicity': ['PopulationGroup'],\n",
       "  'time': ['TemporalConcept'],\n",
       "  'period': ['TemporalConcept']},\n",
       " 121: {'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'cognitive decline': ['DiseaseOrSyndrome'],\n",
       "  'motor function': ['BiologicFunction'],\n",
       "  'Psychosis': ['DiseaseOrSyndrome']},\n",
       " 122: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'pathophysiology': ['PathologicFunction'],\n",
       "  'brain inflammation': ['PathologicFunction', 'DiseaseOrSyndrome'],\n",
       "  'biomarkers': ['ClinicalAttribute'],\n",
       "  'diagnosis': ['HealthCareActivity'],\n",
       "  'saliva': ['BodyPartOrganOrOrganComponent']},\n",
       " 123: {'central nervous systems': ['BodyPartOrganOrOrganComponent'],\n",
       "  'severe': ['Finding'],\n",
       "  'morbidity': ['Finding'],\n",
       "  'drugs': ['PharmacologicSubstance'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome']},\n",
       " 124: {'RhoA': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome']},\n",
       " 125: {'Mitochondria': ['CellComponent'],\n",
       "  'energy metabolism': ['BiologicFunction'],\n",
       "  'encoding': ['BiologicFunction'],\n",
       "  'mitochondrial proteins': ['AminoAcidPeptideOrProtein'],\n",
       "  'mitochondria': ['CellComponent'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome']},\n",
       " 126: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'diffusion': ['NaturalPhenomenonOrProcess'],\n",
       "  'severity': ['Finding'],\n",
       "  'neural degeneration': ['CellOrMolecularDysfunction'],\n",
       "  'MRI': ['HealthCareActivity'],\n",
       "  'DWI': ['HealthCareActivity']},\n",
       " 127: {'mitochondria': ['CellComponent'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'accumulation': ['Finding']},\n",
       " 128: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'severity': ['Finding'],\n",
       "  'relief': ['Finding']},\n",
       " 129: {\"Alzheimer's\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'copper': ['Chemical',\n",
       "   'PharmacologicSubstance',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'iron': ['Chemical',\n",
       "   'PharmacologicSubstance',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'manganese': ['Chemical', 'BiologicallyActiveSubstance'],\n",
       "  'zinc': ['Chemical',\n",
       "   'PharmacologicSubstance',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'CNS': ['BodyPartOrganOrOrganComponent'],\n",
       "  'learning': ['BiologicFunction'],\n",
       "  'homeostasis': ['BiologicFunction']},\n",
       " 130: {'protein homeostasis': ['CellFunction'],\n",
       "  'inhibition': ['CellFunction'],\n",
       "  'accumulation': ['Finding'],\n",
       "  'ubiquitinated proteins': ['AminoAcidPeptideOrProtein'],\n",
       "  'research': ['ResearchActivity'],\n",
       "  'reduction': ['Finding'],\n",
       "  'tau': ['AminoAcidPeptideOrProtein'],\n",
       "  'huntingtin': ['AminoAcidPeptideOrProtein'],\n",
       "  'effectors': ['AminoAcidPeptideOrProtein'],\n",
       "  'relationships': ['Finding']},\n",
       " 131: {'individuals': ['PopulationGroup'],\n",
       "  'dysfunction': ['PathologicFunction'],\n",
       "  'protein homeostasis': ['CellFunction'],\n",
       "  'Administration': ['HealthCareActivity'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'development': ['BiologicFunction'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'frontotemporal lobar degeneration': ['DiseaseOrSyndrome']},\n",
       " 132: {'MicroRNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'miRNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'cell growth': ['CellFunction'],\n",
       "  'cell proliferation': ['CellFunction'],\n",
       "  'embryonic development': ['BiologicFunction'],\n",
       "  'mammalian': ['Eukaryote'],\n",
       "  'microRNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance']},\n",
       " 133: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'rare': ['TemporalConcept'],\n",
       "  'drugs': ['PharmacologicSubstance'],\n",
       "  'tetrabenazine': ['Chemical', 'PharmacologicSubstance']},\n",
       " 134: {'frequent': ['TemporalConcept'],\n",
       "  'Parkinson disease': ['DiseaseOrSyndrome'],\n",
       "  'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'drowsiness': ['Finding', 'DiseaseOrSyndrome']},\n",
       " 135: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'autosomal dominant disorder': ['DiseaseOrSyndrome'],\n",
       "  'mice': ['Eukaryote'],\n",
       "  'pig': ['Eukaryote'],\n",
       "  'sheep': ['Eukaryote'],\n",
       "  'monkeys': ['Eukaryote'],\n",
       "  'pathways': ['CellFunction'],\n",
       "  'Drosophila': ['Eukaryote'],\n",
       "  'model': ['ResearchActivity']},\n",
       " 136: {'mood': ['BiologicFunction'],\n",
       "  'depression': ['DiseaseOrSyndrome'],\n",
       "  'study': ['ResearchActivity'],\n",
       "  'basal ganglia': ['BodyPartOrganOrOrganComponent'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'TMS': ['Finding'],\n",
       "  'Changes': ['Finding']},\n",
       " 137: {'motor function': ['BiologicFunction'],\n",
       "  'age': ['OrganismAttribute'],\n",
       "  'risk factor': ['Finding'],\n",
       "  'polyunsaturated fatty acids': ['Chemical'],\n",
       "  'PUFA': ['Chemical'],\n",
       "  'oxidants': ['Chemical',\n",
       "   'PharmacologicSubstance',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'elevation': ['Finding'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome']},\n",
       " 138: {'biochemical pathways': ['CellFunction'],\n",
       "  'pathways': ['CellFunction'],\n",
       "  'knowledge': ['Finding'],\n",
       "  'metabolic': ['CellFunction']},\n",
       " 139: {'prion protein': ['AminoAcidPeptideOrProtein'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'glia': ['Cell'],\n",
       "  'culture system': ['ResearchActivity'],\n",
       "  'in vivo': ['ResearchActivity']},\n",
       " 140: {'severe': ['Finding'],\n",
       "  'MicroRNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'miRNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'cancer': ['DiseaseOrSyndrome'],\n",
       "  'reactive oxygen species': ['BiologicallyActiveSubstance']},\n",
       " 141: {'transcript': ['NucleicAcidNucleosideOrNucleotide'],\n",
       "  'mRNA': ['NucleicAcidNucleosideOrNucleotide'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'neuronal degeneration': ['CellOrMolecularDysfunction'],\n",
       "  'evaluation': ['HealthCareActivity', 'ResearchActivity'],\n",
       "  'treatments': ['HealthCareActivity']},\n",
       " 142: {'knowledge': ['Finding'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent']},\n",
       " 143: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'brain pathologies': ['PathologicFunction'],\n",
       "  'aggregation': ['CellFunction'],\n",
       "  'cardiac': ['BodyPartOrganOrOrganComponent'],\n",
       "  'involvement': ['Finding'],\n",
       "  'studies': ['ResearchActivity'],\n",
       "  'cardiac function': ['BiologicFunction'],\n",
       "  'current': ['TemporalConcept'],\n",
       "  'knowledge': ['Finding'],\n",
       "  'electrocardiogram': ['HealthCareActivity'],\n",
       "  'cardiac electrophysiology': ['HealthCareActivity']},\n",
       " 144: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'Microglial': ['Cell'],\n",
       "  'evidence': ['Finding'],\n",
       "  'cerebrospinal fluid': ['BodyPartOrganOrOrganComponent'],\n",
       "  'monitor': ['HealthCareActivity']},\n",
       " 145: {'Antisense oligonucleotides': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance',\n",
       "   'Chemical'],\n",
       "  'medicines': ['PharmacologicSubstance'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'nusinersen': ['Chemical', 'PharmacologicSubstance']},\n",
       " 146: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'huntingtin gene': ['GeneOrGenome'],\n",
       "  'CNS': ['BodyPartOrganOrOrganComponent'],\n",
       "  'inflammation': ['PathologicFunction'],\n",
       "  'IL-6': ['GeneOrGenome'],\n",
       "  'Ntrk2': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance']},\n",
       " 147: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'transcription factor nuclear factor erythroid 2-related factor 2': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'NRF2': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'biogenesis': ['CellFunction']},\n",
       " 148: {'Orexin': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'neurotransmitter': ['BiologicallyActiveSubstance'],\n",
       "  'sleep/wake cycle': ['BiologicFunction'],\n",
       "  'attention': ['BiologicFunction'],\n",
       "  'involvement': ['Finding'],\n",
       "  'orexin': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  \"Alzheimer's\": ['DiseaseOrSyndrome'],\n",
       "  'Amyotrophic lateral sclerosis': ['DiseaseOrSyndrome']},\n",
       " 149: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'infrequent': ['TemporalConcept'],\n",
       "  'girl': ['PopulationGroup'],\n",
       "  'age': ['OrganismAttribute'],\n",
       "  'Huntingtin gene': ['GeneOrGenome']},\n",
       " 150: {'nicotinamide adenine dinucleotide': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance',\n",
       "   'PharmacologicSubstance'],\n",
       "  'nicotinamide riboside': ['BiologicallyActiveSubstance'],\n",
       "  'Sirtuin 3': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'ROS': ['Chemical', 'BiologicallyActiveSubstance'],\n",
       "  'bioenergetics': ['BiologicFunction']},\n",
       " 151: {'pathophysiology': ['PathologicFunction'],\n",
       "  'anatomy': ['BiomedicalOccupationOrDiscipline'],\n",
       "  'genome': ['GeneOrGenome'],\n",
       "  'animal': ['Eukaryote'],\n",
       "  'research': ['ResearchActivity'],\n",
       "  \"Alzheimer's Disease\": ['DiseaseOrSyndrome']},\n",
       " 152: {'pathogenicity': ['OrganismAttribute'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'huntingtin': ['GeneOrGenome', 'AminoAcidPeptideOrProtein'],\n",
       "  'toxicity': ['InjuryOrPoisoning'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'cellular response': ['CellFunction']},\n",
       " 153: {'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'brain tissue': ['BodyPartOrganOrOrganComponent'],\n",
       "  'study': ['ResearchActivity'],\n",
       "  'pathophysiology': ['PathologicFunction'],\n",
       "  'biomarkers': ['ClinicalAttribute'],\n",
       "  'blood-brain barrier': ['BodyPartOrganOrOrganComponent'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'CNS': ['BodyPartOrganOrOrganComponent']},\n",
       " 154: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'autosomal dominant disorder': ['DiseaseOrSyndrome']},\n",
       " 155: {'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'study': ['ResearchActivity'],\n",
       "  'quality of life': ['Finding']},\n",
       " 156: {'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'CNS': ['BodyPartOrganOrOrganComponent'],\n",
       "  'human': ['Eukaryote'],\n",
       "  'preservation': ['HealthCareActivity'],\n",
       "  'chemopreventive': ['PharmacologicSubstance'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome']},\n",
       " 157: {'Deutetrabenazine': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'hydrogen': ['Chemical'],\n",
       "  'half-life': ['TemporalConcept'],\n",
       "  'placebo': ['HealthCareActivity'],\n",
       "  'tetrabenazine': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'review': ['HealthCareActivity']},\n",
       " 158: {'MicroRNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'miRNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'health': ['HealthCareActivity'],\n",
       "  'human': ['Eukaryote'],\n",
       "  'review': ['ResearchActivity']},\n",
       " 160: {'class-II major histocompatibility complex': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'neuroprotective': ['CellFunction'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome']},\n",
       " 161: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'abnormal': ['Finding'],\n",
       "  'damage': ['InjuryOrPoisoning'],\n",
       "  'PET': ['HealthCareActivity'],\n",
       "  'magnetic resonance imaging': ['HealthCareActivity'],\n",
       "  'MRI': ['HealthCareActivity'],\n",
       "  'clinical trials': ['ResearchActivity'],\n",
       "  'research': ['ResearchActivity']},\n",
       " 163: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'public health': ['HealthCareActivity', 'BiomedicalOccupationOrDiscipline'],\n",
       "  'manganese': ['Chemical', 'BiologicallyActiveSubstance'],\n",
       "  'Mn': ['BiologicallyActiveSubstance', 'Chemical'],\n",
       "  'development': ['CellFunction', 'BiologicFunction']},\n",
       " 165: {'Gangliosides': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'CNS': ['BodyPartOrganOrOrganComponent'],\n",
       "  'BSE': ['DiseaseOrSyndrome'],\n",
       "  'gangliosides': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'pre-clinical': ['ResearchActivity'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'randomized controlled trials': ['ResearchActivity'],\n",
       "  'no difference': ['Finding'],\n",
       "  'survival': ['CellFunction'],\n",
       "  'placebo': ['HealthCareActivity'],\n",
       "  'GM1': ['Chemical', 'PharmacologicSubstance', 'BiologicallyActiveSubstance'],\n",
       "  'phosphorylation': ['CellFunction'],\n",
       "  'reduction': ['Finding', 'NaturalPhenomenonOrProcess'],\n",
       "  'endoplasmic reticulum': ['CellComponent'],\n",
       "  'recovery': ['BiologicFunction'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'in vitro': ['ResearchActivity']},\n",
       " 171: {'nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'schizophrenia': ['DiseaseOrSyndrome'],\n",
       "  'depression': ['DiseaseOrSyndrome'],\n",
       "  'study': ['ResearchActivity'],\n",
       "  'treatment': ['HealthCareActivity']},\n",
       " 175: {'autoimmunity': ['PathologicFunction'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome']},\n",
       " 176: {'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'huntingtin gene': ['GeneOrGenome'],\n",
       "  'movement disorder': ['DiseaseOrSyndrome'],\n",
       "  'cognitive decline': ['DiseaseOrSyndrome'],\n",
       "  'patient': ['PatientOrDisabledGroup'],\n",
       "  'hypothermia': ['InjuryOrPoisoning']},\n",
       " 179: {'research': ['ResearchActivity']},\n",
       " 182: {'severe': ['Finding'],\n",
       "  'autosomal dominant': ['CellFunction'],\n",
       "  'mutation': ['CellOrMolecularDysfunction'],\n",
       "  'Huntingtin gene': ['GeneOrGenome'],\n",
       "  'death': ['CellFunction'],\n",
       "  'expressed': ['CellFunction'],\n",
       "  'protein degradation': ['CellFunction'],\n",
       "  'review': ['ResearchActivity']},\n",
       " 183: {'cognitive decline': ['DiseaseOrSyndrome'],\n",
       "  'development': ['CellFunction'],\n",
       "  'study': ['ResearchActivity'],\n",
       "  'MSH3': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'BDNF': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'COMT': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'MAPT': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'MSH3 gene': ['GeneOrGenome'],\n",
       "  'evidence': ['Finding']},\n",
       " 184: {'cognitive decline': ['DiseaseOrSyndrome'],\n",
       "  'ataxia': ['SignOrSymptom'],\n",
       "  'dystonia': ['SignOrSymptom', 'DiseaseOrSyndrome']},\n",
       " 191: {'disorders': ['DiseaseOrSyndrome'],\n",
       "  'elderly': ['PopulationGroup'],\n",
       "  'development': ['CellFunction', 'BiologicFunction'],\n",
       "  'sporadic': ['TemporalConcept'],\n",
       "  'Parkinson disease': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'Alzheimer disease': ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'spinocerebellar ataxias': ['DiseaseOrSyndrome'],\n",
       "  'SCA': ['DiseaseOrSyndrome']},\n",
       " 197: {'long-term': ['TemporalConcept'], 'HD': ['DiseaseOrSyndrome']},\n",
       " 201: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'rare': ['TemporalConcept'],\n",
       "  'Cognitive impairment': ['DiseaseOrSyndrome'],\n",
       "  'patient': ['PatientOrDisabledGroup'],\n",
       "  'cognitive impairment': ['DiseaseOrSyndrome'],\n",
       "  'imaging': ['HealthCareActivity'],\n",
       "  'molecular genetic testing': ['BiomedicalOccupationOrDiscipline'],\n",
       "  'diagnosis': ['HealthCareActivity']},\n",
       " 202: {'movement disorder': ['DiseaseOrSyndrome'],\n",
       "  'humans': ['Eukaryote'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'Amyotrophic Lateral Sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'human': ['Eukaryote'],\n",
       "  'PCA': ['ResearchActivity']},\n",
       " 203: {'zebrafish': ['Eukaryote'],\n",
       "  'human': ['Eukaryote'],\n",
       "  'neuropathology': ['PathologicFunction', 'BiomedicalOccupationOrDiscipline'],\n",
       "  'drug screening': ['ResearchActivity'],\n",
       "  'review': ['HealthCareActivity']},\n",
       " 206: {'chronic': ['TemporalConcept'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'lipid': ['Chemical'],\n",
       "  'programmed cell death': ['CellFunction'],\n",
       "  'occurrence': ['TemporalConcept'],\n",
       "  'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'research': ['ResearchActivity']},\n",
       " 208: {'NPY': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'mammalian': ['Eukaryote'],\n",
       "  'Amyotrophic Lateral Sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'Machado-Joseph disease': ['DiseaseOrSyndrome']},\n",
       " 209: {'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'huntingtin gene': ['GeneOrGenome'],\n",
       "  'chromosome 4': ['CellComponent'],\n",
       "  'treatments': ['HealthCareActivity']},\n",
       " 213: {'HD': ['DiseaseOrSyndrome'], 'treatment': ['ResearchActivity']},\n",
       " 214: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'pathophysiology': ['PathologicFunction'],\n",
       "  '2-arachidonoylglycerol': ['Chemical']},\n",
       " 224: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'study': ['ResearchActivity'],\n",
       "  'Participants': ['PopulationGroup'],\n",
       "  'symptom severity': ['Finding']},\n",
       " 228: {'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'deutetrabenazine': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'treatment': ['ResearchActivity'],\n",
       "  'mutated huntingtin protein': ['AminoAcidPeptideOrProtein'],\n",
       "  'measure': ['HealthCareActivity']},\n",
       " 230: {'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'degeneration': ['PathologicFunction']},\n",
       " 232: {'Dimethyl fumarate': ['Chemical',\n",
       "   'PharmacologicSubstance',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'DMF': ['Chemical', 'PharmacologicSubstance', 'BiologicallyActiveSubstance'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'psoriasis': ['DiseaseOrSyndrome'],\n",
       "  'monomethyl fumarate': ['Chemical'],\n",
       "  'MMF': ['Chemical'],\n",
       "  'fumarates': ['Chemical'],\n",
       "  'inhibition': ['CellFunction'],\n",
       "  'agonist': ['BiologicallyActiveSubstance'],\n",
       "  'PD': ['DiseaseOrSyndrome']},\n",
       " 233: {'research': ['ResearchActivity'],\n",
       "  'human': ['Eukaryote'],\n",
       "  'health': ['HealthCareActivity'],\n",
       "  'bioactive compounds': ['BiologicallyActiveSubstance'],\n",
       "  'caffeic acid': ['Chemical', 'BiologicallyActiveSubstance'],\n",
       "  'alkaloids': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'polyphenols': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'studies': ['ResearchActivity'],\n",
       "  'dysfunction': ['PathologicFunction'],\n",
       "  'time': ['TemporalConcept'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'MS': ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome']},\n",
       " 234: {'epigenetics': ['CellFunction', 'BiomedicalOccupationOrDiscipline'],\n",
       "  'DNA methylation': ['CellFunction'],\n",
       "  'histones': ['AminoAcidPeptideOrProtein'],\n",
       "  'long non-coding RNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'Amyotrophic Lateral Sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'lifespan': ['TemporalConcept'],\n",
       "  'treatment': ['HealthCareActivity']},\n",
       " 235: {'essential oils': ['Chemical'],\n",
       "  'effectiveness': ['Finding'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'study': ['ResearchActivity']},\n",
       " 245: {'HD': ['DiseaseOrSyndrome']},\n",
       " 253: {'neurotransmitter': ['BiologicallyActiveSubstance'],\n",
       "  'murine': ['Eukaryote'],\n",
       "  'studies': ['ResearchActivity'],\n",
       "  'kinetics': ['NaturalPhenomenonOrProcess'],\n",
       "  'mutation': ['CellOrMolecularDysfunction'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'mutated huntingtin': ['AminoAcidPeptideOrProtein'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'huntingtin': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance']},\n",
       " 255: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'NLRP3': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'inflammasome': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'expressed': ['CellFunction'],\n",
       "  'toxicity': ['InjuryOrPoisoning'],\n",
       "  'pathology': ['PathologicFunction', 'BiomedicalOccupationOrDiscipline'],\n",
       "  'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'microglial': ['Cell'],\n",
       "  'ROS': ['BiologicallyActiveSubstance']},\n",
       " 256: {'HD': ['DiseaseOrSyndrome'], 'frequent': ['TemporalConcept']},\n",
       " 258: {'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'CNS': ['BodyPartOrganOrOrganComponent'],\n",
       "  'plaques': ['AnatomicalAbnormality'],\n",
       "  'development': ['CellFunction', 'BiologicFunction'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome']},\n",
       " 259: {'Iron': ['Chemical',\n",
       "   'PharmacologicSubstance',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'pathology': ['PathologicFunction', 'BiomedicalOccupationOrDiscipline'],\n",
       "  'development': ['CellFunction', 'BiologicFunction'],\n",
       "  'iron': ['Chemical',\n",
       "   'PharmacologicSubstance',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'synthesis': ['BiologicFunction'],\n",
       "  'studies': ['ResearchActivity'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'spinocerebellar ataxia': ['DiseaseOrSyndrome']},\n",
       " 261: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'Inflammasomes': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'multiprotein complexes': ['AminoAcidPeptideOrProtein'],\n",
       "  'inflammasomes': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'caspases': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'interleukins': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'caspase-1': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'cleavage': ['CellFunction'],\n",
       "  'IL18': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'NLRP3 inflammasome complex': ['CellComponent']},\n",
       " 262: {'autosomal dominant': ['CellFunction'],\n",
       "  'huntingtin gene': ['GeneOrGenome'],\n",
       "  'huntingtin': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'Aggregation': ['CellFunction']},\n",
       " 263: {'Amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'toxic proteins': ['AminoAcidPeptideOrProtein']},\n",
       " 264: {'Identification': ['BiologicFunction'],\n",
       "  'biomarkers': ['ClinicalAttribute'],\n",
       "  'biomarker': ['ClinicalAttribute'],\n",
       "  'decrease': ['Finding']},\n",
       " 265: {'MicroRNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'gene expression': ['CellFunction'],\n",
       "  'neuronal differentiation': ['CellFunction'],\n",
       "  'survival': ['CellFunction'],\n",
       "  'activity': ['CellFunction'],\n",
       "  'microRNA': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'vulnerability': ['Finding'],\n",
       "  'development': ['CellFunction', 'BiologicFunction'],\n",
       "  'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'microRNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance']},\n",
       " 266: {'world': ['PopulationGroup'],\n",
       "  'treatments': ['HealthCareActivity'],\n",
       "  'symptoms': ['SignOrSymptom'],\n",
       "  'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'attention': ['BiologicFunction'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'drug': ['PharmacologicSubstance'],\n",
       "  'anxiety': ['DiseaseOrSyndrome'],\n",
       "  'severity': ['ClinicalAttribute']},\n",
       " 267: {'dysfunction': ['PathologicFunction'],\n",
       "  'nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'biomarkers': ['ClinicalAttribute'],\n",
       "  'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'treatment response': ['ClinicalAttribute'],\n",
       "  'prognosis': ['HealthCareActivity'],\n",
       "  'proteins': ['AminoAcidPeptideOrProtein'],\n",
       "  'nucleic acids': ['NucleicAcidNucleosideOrNucleotide'],\n",
       "  'lipids': ['Chemical'],\n",
       "  'disease progression': ['PathologicFunction'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'miRNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance']},\n",
       " 268: {'Acetylation': ['CellFunction'],\n",
       "  'post-translational modification': ['CellFunction'],\n",
       "  'histone': ['AminoAcidPeptideOrProtein'],\n",
       "  'metabolism': ['BiologicFunction'],\n",
       "  'memory': ['BiologicFunction'],\n",
       "  'Homeostasis': ['BiologicFunction'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'Huntingtin protein': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance']},\n",
       " 269: {'frequency': ['TemporalConcept'],\n",
       "  'tau': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'proteins': ['AminoAcidPeptideOrProtein'],\n",
       "  'trio': ['AminoAcidPeptideOrProtein']},\n",
       " 270: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'treatments': ['HealthCareActivity'],\n",
       "  'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'age': ['OrganismAttribute'],\n",
       "  'risk': ['Finding']},\n",
       " 271: {'proteins': ['AminoAcidPeptideOrProtein'],\n",
       "  'human': ['Eukaryote'],\n",
       "  'transglutaminases': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'expressed': ['CellFunction'],\n",
       "  'CNS': ['BodyPartOrganOrOrganComponent'],\n",
       "  'study': ['ResearchActivity']},\n",
       " 272: {'Huntington Disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'pressure': ['BiologicFunction'],\n",
       "  'classification': ['ResearchActivity']},\n",
       " 273: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'development': ['CellFunction', 'BiologicFunction'],\n",
       "  'treatments': ['HealthCareActivity'],\n",
       "  'time': ['TemporalConcept'],\n",
       "  'drugs': ['PharmacologicSubstance']},\n",
       " 275: {'dysfunctions': ['PathologicFunction'],\n",
       "  'toxic': ['InjuryOrPoisoning'],\n",
       "  'inclusion bodies': ['CellComponent'],\n",
       "  'impairment': ['Finding'],\n",
       "  'human': ['Eukaryote'],\n",
       "  'unfolded protein response': ['CellFunction'],\n",
       "  'pathophysiology': ['PathologicFunction'],\n",
       "  \"Alzheimer's\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'exacerbated': ['Finding'],\n",
       "  'deteriorating': ['Finding'],\n",
       "  'components': ['CellComponent'],\n",
       "  'worsen': ['Finding'],\n",
       "  'neurodegenerative': ['CellOrMolecularDysfunction'],\n",
       "  'therapeutics': ['HealthCareActivity'],\n",
       "  'development': ['CellFunction', 'BiologicFunction']},\n",
       " 276: {'PD': ['DiseaseOrSyndrome'],\n",
       "  'Amyotrophic Lateral Sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome']},\n",
       " 277: {'degeneration': ['PathologicFunction'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome']},\n",
       " 278: {'DNA mismatch repair': ['CellFunction'],\n",
       "  'MMR': ['CellFunction'],\n",
       "  'cellular response': ['CellFunction'],\n",
       "  'pathway': ['CellFunction']},\n",
       " 279: {'mitochondria': ['CellComponent']},\n",
       " 281: {'rare': ['TemporalConcept'],\n",
       "  'severe': ['Finding'],\n",
       "  'grafting': ['BodyPartOrganOrOrganComponent', 'HealthCareActivity'],\n",
       "  'early stage': ['TemporalConcept'],\n",
       "  'period': ['TemporalConcept'],\n",
       "  'month': ['TemporalConcept'],\n",
       "  'transplanted': ['HealthCareActivity'],\n",
       "  'time': ['TemporalConcept'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'delayed': ['TemporalConcept'],\n",
       "  'rate': ['TemporalConcept'],\n",
       "  'transplantation': ['HealthCareActivity'],\n",
       "  'procedures': ['HealthCareActivity', 'TemporalConcept']},\n",
       " 282: {'homeostasis': ['BiologicFunction'],\n",
       "  'intercellular bridges': ['CellComponent'],\n",
       "  'pharmacology': ['BiomedicalOccupationOrDiscipline'],\n",
       "  'alpha-synuclein': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'skeleton': ['BodyPartOrganOrOrganComponent']},\n",
       " 291: {'Huntington Disease': ['DiseaseOrSyndrome'],\n",
       "  'death': ['BiologicFunction'],\n",
       "  'problem': ['Finding'],\n",
       "  'decision-making': ['BiologicFunction']},\n",
       " 292: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'expression': ['CellFunction'],\n",
       "  'mutant huntingtin protein': ['AminoAcidPeptideOrProtein'],\n",
       "  'intracellular': ['CellComponent'],\n",
       "  'inclusions': ['CellComponent'],\n",
       "  'cell types': ['Cell'],\n",
       "  'urine': ['BodyPartOrganOrOrganComponent'],\n",
       "  'saliva': ['BodyPartOrganOrOrganComponent'],\n",
       "  'cerebrospinal fluid': ['BodyPartOrganOrOrganComponent'],\n",
       "  'toxic': ['InjuryOrPoisoning'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'review': ['HealthCareActivity']},\n",
       " 296: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'encoding': ['BiologicFunction'],\n",
       "  'huntingtin protein': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'IL-6': ['GeneOrGenome'],\n",
       "  'IL-4': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'decrease': ['Finding'],\n",
       "  'IL-5': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance']},\n",
       " 301: {'mutation': ['CellOrMolecularDysfunction'],\n",
       "  'huntingtin': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'metabolism': ['CellFunction', 'BiologicFunction']},\n",
       " 307: {'cellular homeostasis': ['CellFunction'],\n",
       "  'evidence': ['Finding'],\n",
       "  'CNS': ['BodyPartOrganOrOrganComponent'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'plasma': ['BodyPartOrganOrOrganComponent']},\n",
       " 313: {'dystonia': ['SignOrSymptom'],\n",
       "  'mutated protein': ['AminoAcidPeptideOrProtein'],\n",
       "  'attention': ['BiologicFunction'],\n",
       "  'relationship': ['Finding'],\n",
       "  'microRNAs': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'transcription factors': ['BiologicallyActiveSubstance'],\n",
       "  'microRNA': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'brain-derived neurotrophic factor genes': ['GeneOrGenome'],\n",
       "  'homeobox': ['GeneOrGenome', 'AminoAcidPeptideOrProtein']},\n",
       " 314: {'ChAc': ['DiseaseOrSyndrome']},\n",
       " 318: {'neuropathology': ['PathologicFunction'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'study': ['ResearchActivity'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'transcripts': ['NucleicAcidNucleosideOrNucleotide']},\n",
       " 329: {'Small heat shock proteins': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'chaperones': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'HSPB1': ['GeneOrGenome', 'AminoAcidPeptideOrProtein'],\n",
       "  'HSPB8': ['GeneOrGenome', 'AminoAcidPeptideOrProtein']},\n",
       " 340: {'Curcumin': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'NF-kappa B': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'signal transducer and activator of transcription 3': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'STAT3': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'nuclear factor erythroid 2-related factor 2': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'Nrf2': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'cyclooxygenases-2': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'autosomal dominant disorder': ['DiseaseOrSyndrome'],\n",
       "  'gene expression': ['CellFunction'],\n",
       "  'adenine': ['BiologicallyActiveSubstance'],\n",
       "  'side effects': ['PathologicFunction']},\n",
       " 343: {'humans': ['Eukaryote'],\n",
       "  'in vivo': ['ResearchActivity'],\n",
       "  'review': ['HealthCareActivity']},\n",
       " 350: {'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'CNS': ['BodyPartOrganOrOrganComponent'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'MS': ['DiseaseOrSyndrome'],\n",
       "  'Colony stimulating factor 1-receptor': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'CSF-1R': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'Research': ['ResearchActivity'],\n",
       "  'humans': ['Eukaryote']},\n",
       " 365: {'lifespan': ['TemporalConcept'],\n",
       "  'neuronal degeneration': ['CellOrMolecularDysfunction'],\n",
       "  'degenerated': ['PathologicFunction'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'CNS': ['BodyPartOrganOrOrganComponent'],\n",
       "  'reprogramming': ['CellFunction'],\n",
       "  'Human': ['Eukaryote'],\n",
       "  'in vivo': ['ResearchActivity'],\n",
       "  'HD': ['DiseaseOrSyndrome']},\n",
       " 366: {'cell division': ['CellFunction'],\n",
       "  'DNA damage repair': ['CellFunction'],\n",
       "  'stress responses': ['CellFunction', 'BiologicFunction'],\n",
       "  'attention': ['BiologicFunction'],\n",
       "  'individuals': ['PopulationGroup'],\n",
       "  'risk': ['Finding'],\n",
       "  'treatments': ['HealthCareActivity'],\n",
       "  'cell': ['Cell'],\n",
       "  'normal': ['Finding'],\n",
       "  'reports': ['HealthCareActivity']},\n",
       " 368: {'Huntingtin-associated protein 1': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'HAP1': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'expressed': ['CellFunction'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'SBMA': ['DiseaseOrSyndrome'],\n",
       "  'spinocerebellar ataxia': ['DiseaseOrSyndrome'],\n",
       "  'SCA': ['DiseaseOrSyndrome']},\n",
       " 379: {'Vitamins': ['BiologicallyActiveSubstance'],\n",
       "  'micronutrients': ['BiologicallyActiveSubstance'],\n",
       "  'immune system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'infection': ['PathologicFunction'],\n",
       "  'synuclein': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'Depression': ['DiseaseOrSyndrome'],\n",
       "  'vitamin-C': ['BiologicallyActiveSubstance'],\n",
       "  'vitamin D': ['BiologicallyActiveSubstance'],\n",
       "  'acetylcholine': ['Chemical',\n",
       "   'PharmacologicSubstance',\n",
       "   'BiologicallyActiveSubstance']},\n",
       " 380: {'BDNF': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'survival': ['CellFunction'],\n",
       "  'maturation': ['BiologicFunction'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'cognitive impairment': ['DiseaseOrSyndrome']},\n",
       " 384: {'research': ['ResearchActivity'],\n",
       "  'diagnosis': ['HealthCareActivity'],\n",
       "  \"Alzheimer's\": ['DiseaseOrSyndrome'],\n",
       "  'Amyotrophic Lateral Sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'analysis': ['ResearchActivity'],\n",
       "  'sleep disorders': ['DiseaseOrSyndrome'],\n",
       "  'current': ['TemporalConcept']},\n",
       " 397: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'humans': ['Eukaryote'],\n",
       "  'pathologies': ['BiomedicalOccupationOrDiscipline'],\n",
       "  'pathway': ['CellFunction'],\n",
       "  'dysfunctional': ['PathologicFunction'],\n",
       "  'mitochondria': ['CellComponent'],\n",
       "  'pathways': ['CellFunction'],\n",
       "  'treatment': ['HealthCareActivity']},\n",
       " 402: {'pain': ['SignOrSymptom'],\n",
       "  'death': ['BiologicFunction'],\n",
       "  'human': ['Eukaryote'],\n",
       "  'gene variants': ['GeneOrGenome'],\n",
       "  'research': ['ResearchActivity'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'Amyotrophic Lateral Sclerosis': ['DiseaseOrSyndrome']},\n",
       " 412: {'retinal': ['BodyPartOrganOrOrganComponent'],\n",
       "  'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome']},\n",
       " 436: {'HD': ['DiseaseOrSyndrome']},\n",
       " 437: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'management': ['HealthCareActivity'],\n",
       "  'treatments': ['HealthCareActivity'],\n",
       "  'randomized controlled trials': ['ResearchActivity'],\n",
       "  'interventions': ['HealthCareActivity'],\n",
       "  'tetrabenazine': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'depression': ['DiseaseOrSyndrome'],\n",
       "  'physiotherapy': ['HealthCareActivity']},\n",
       " 438: {'Neurotrophins': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'BDNF': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'growth': ['BiologicFunction'],\n",
       "  'receptor': ['BiologicallyActiveSubstance']},\n",
       " 439: {'Rhodopsin': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'G-protein coupled receptor': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'biomarker': ['ClinicalAttribute'],\n",
       "  'early detection': ['HealthCareActivity'],\n",
       "  'monitoring': ['HealthCareActivity'],\n",
       "  'rhodopsin': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'technique': ['ResearchActivity']},\n",
       " 445: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'Research': ['ResearchActivity'],\n",
       "  'neurotrophic factors': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance']},\n",
       " 454: {'accumulation': ['Finding'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'toxicity': ['InjuryOrPoisoning'],\n",
       "  'nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'cell': ['Cell'],\n",
       "  'Molecular chaperones': ['AminoAcidPeptideOrProtein'],\n",
       "  'damage': ['InjuryOrPoisoning'],\n",
       "  'age': ['OrganismAttribute'],\n",
       "  'studies': ['ResearchActivity'],\n",
       "  'chaperones': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'chaperone': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance']},\n",
       " 460: {'group': ['PopulationGroup'],\n",
       "  'period': ['TemporalConcept'],\n",
       "  'inflammation': ['PathologicFunction'],\n",
       "  'accumulation': ['Finding'],\n",
       "  'toxic proteins': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'manganese': ['Chemical', 'BiologicallyActiveSubstance'],\n",
       "  'Mn': ['BiologicallyActiveSubstance', 'Chemical'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'serotonin': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'alpha-synuclein': ['BiologicallyActiveSubstance']},\n",
       " 469: {'retinal': ['BodyPartOrganOrOrganComponent'],\n",
       "  'biomarker': ['ClinicalAttribute'],\n",
       "  'HD': ['DiseaseOrSyndrome']},\n",
       " 471: {'problem': ['Finding'],\n",
       "  'Alzheimer disease': ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'Parkinson disease': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'HD protein': ['AminoAcidPeptideOrProtein'],\n",
       "  'observations': ['Finding']},\n",
       " 472: {'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'exon': ['NucleicAcidNucleosideOrNucleotide'],\n",
       "  'huntingtin gene': ['GeneOrGenome'],\n",
       "  'dysfunction': ['PathologicFunction'],\n",
       "  'synaptic transmission': ['CellFunction'],\n",
       "  'glutamate': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'gamma-aminobutyric acid': ['AminoAcidPeptideOrProtein'],\n",
       "  'dopamine': ['BiologicallyActiveSubstance', 'PharmacologicSubstance'],\n",
       "  'adenosine': ['PharmacologicSubstance', 'BiologicallyActiveSubstance'],\n",
       "  'endocannabinoids': ['BiologicallyActiveSubstance'],\n",
       "  'brain-derived neurotrophic factor': ['BiologicallyActiveSubstance'],\n",
       "  'signaling pathways': ['CellFunction'],\n",
       "  'hippocampus': ['BodyPartOrganOrOrganComponent']},\n",
       " 473: {'symptoms': ['SignOrSymptom'],\n",
       "  'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'death': ['BiologicFunction'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'patient': ['PatientOrDisabledGroup'],\n",
       "  'family': ['PopulationGroup'],\n",
       "  'current': ['TemporalConcept'],\n",
       "  'knowledge': ['Finding'],\n",
       "  'group': ['PopulationGroup'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'reviewed': ['HealthCareActivity']},\n",
       " 486: {'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'cognitive decline': ['DiseaseOrSyndrome'],\n",
       "  'expressed': ['CellFunction']},\n",
       " 500: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'treatments': ['HealthCareActivity'],\n",
       "  'murine': ['Eukaryote'],\n",
       "  'clinical trials': ['ResearchActivity'],\n",
       "  'purinergic P2X7 receptors': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'inflammation': ['PathologicFunction'],\n",
       "  'studies': ['ResearchActivity'],\n",
       "  'increasing': ['Finding']},\n",
       " 503: {'research': ['ResearchActivity'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'in vivo': ['ResearchActivity']},\n",
       " 519: {'cancer': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'gene expression': ['CellFunction'],\n",
       "  'survey': ['HealthCareActivity']},\n",
       " 541: {'Cysteine': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'protein synthesis': ['CellFunction'],\n",
       "  'cysteine': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'protein huntingtin': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'study': ['ResearchActivity']},\n",
       " 548: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'blood-brain barrier': ['BodyPartOrganOrOrganComponent']},\n",
       " 552: {'HD': ['DiseaseOrSyndrome'], 'decrease': ['Finding']},\n",
       " 553: {\"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'AD': ['DiseaseOrSyndrome'],\n",
       "  'PD': ['DiseaseOrSyndrome'],\n",
       "  'symptoms': ['SignOrSymptom'],\n",
       "  'clinical course': ['TemporalConcept'],\n",
       "  'diagnostic criteria': ['Finding'],\n",
       "  'pathology': ['PathologicFunction', 'BiomedicalOccupationOrDiscipline'],\n",
       "  'drug development': ['ResearchActivity'],\n",
       "  'kynurenines': ['BiologicallyActiveSubstance']},\n",
       " 557: {'Proteins': ['AminoAcidPeptideOrProtein'],\n",
       "  \"Alzheimer's\": ['DiseaseOrSyndrome'],\n",
       "  'medicines': ['PharmacologicSubstance'],\n",
       "  'illnesses': ['DiseaseOrSyndrome'],\n",
       "  'symptoms': ['SignOrSymptom'],\n",
       "  'human': ['Eukaryote'],\n",
       "  'protein aggregation': ['CellFunction']},\n",
       " 574: {'nucleic acid': ['NucleicAcidNucleosideOrNucleotide'],\n",
       "  'proteins': ['AminoAcidPeptideOrProtein'],\n",
       "  'speed': ['Chemical'],\n",
       "  'mRNA': ['NucleicAcidNucleosideOrNucleotide'],\n",
       "  'protein synthesis': ['CellFunction'],\n",
       "  'Huntingtin': ['BiologicallyActiveSubstance']},\n",
       " 575: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'mutation': ['CellOrMolecularDysfunction'],\n",
       "  'Huntingtin gene': ['GeneOrGenome'],\n",
       "  'miRNA': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'non-coding RNA': ['NucleicAcidNucleosideOrNucleotide',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'review': ['HealthCareActivity']},\n",
       " 576: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'CAG trinucleotide repeat': ['NucleicAcidNucleosideOrNucleotide'],\n",
       "  'expressed': ['CellFunction'],\n",
       "  'toxicity': ['InjuryOrPoisoning'],\n",
       "  'degeneration': ['PathologicFunction']},\n",
       " 577: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'Huntingtin gene': ['GeneOrGenome'],\n",
       "  'IRF7': ['GeneOrGenome']},\n",
       " 579: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'dysphagia': ['DiseaseOrSyndrome'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'studies': ['ResearchActivity']},\n",
       " 590: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'research': ['ResearchActivity'],\n",
       "  'individuals': ['PopulationGroup']},\n",
       " 597: {'Iron': ['Chemical',\n",
       "   'PharmacologicSubstance',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'homeostasis': ['BiologicFunction'],\n",
       "  'iron': ['Chemical',\n",
       "   'PharmacologicSubstance',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'development': ['CellFunction', 'BiologicFunction']},\n",
       " 602: {\"Alzheimer's\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'pontocerebellar hypoplasia': ['DiseaseOrSyndrome'],\n",
       "  'toxicity': ['InjuryOrPoisoning'],\n",
       "  'demyelination': ['PathologicFunction'],\n",
       "  'human': ['Eukaryote']},\n",
       " 613: {'HD': ['DiseaseOrSyndrome']},\n",
       " 625: {'sleep': ['BiologicFunction'],\n",
       "  'pathophysiology': ['PathologicFunction']},\n",
       " 630: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'risk': ['Finding'],\n",
       "  'arrhythmia': ['DiseaseOrSyndrome'],\n",
       "  'susceptibility': ['Finding'],\n",
       "  'PVCs': ['DiseaseOrSyndrome']},\n",
       " 634: {'impairments': ['Finding'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'Studies': ['ResearchActivity'],\n",
       "  'individuals': ['PopulationGroup'],\n",
       "  'carbon dioxide': ['Chemical', 'BiologicallyActiveSubstance'],\n",
       "  'CO2': ['Chemical', 'BiologicallyActiveSubstance'],\n",
       "  'study': ['ResearchActivity']},\n",
       " 642: {'biomarkers': ['ClinicalAttribute'],\n",
       "  'disease progression': ['PathologicFunction'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'degeneration': ['PathologicFunction']},\n",
       " 662: {'Renin-Angiotensin System': ['BodyPartOrganOrOrganComponent'],\n",
       "  'expressed': ['CellFunction'],\n",
       "  'central nervous system': ['BodyPartOrganOrOrganComponent'],\n",
       "  'ACE2': ['GeneOrGenome']},\n",
       " 671: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'Neurofilament light protein': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'progression': ['PathologicFunction'],\n",
       "  'GFAP': ['GeneOrGenome']},\n",
       " 681: {'disability': ['Finding'],\n",
       "  \"Alzheimer's\": ['DiseaseOrSyndrome'],\n",
       "  'cognitive decline': ['DiseaseOrSyndrome'],\n",
       "  'symptom': ['SignOrSymptom'],\n",
       "  'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'neighbouring': ['PopulationGroup'],\n",
       "  'vesicles': ['CellComponent']},\n",
       " 723: {'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'deteriorating': ['Finding'],\n",
       "  'cognitive impairment': ['DiseaseOrSyndrome'],\n",
       "  'psychosis': ['DiseaseOrSyndrome'],\n",
       "  'Olanzapine': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'Memantine': ['Chemical', 'PharmacologicSubstance'],\n",
       "  'Tetrabenazine': ['Chemical', 'PharmacologicSubstance']},\n",
       " 729: {'Huntington disease': ['DiseaseOrSyndrome'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'huntingtin gene': ['GeneOrGenome'],\n",
       "  'nucleocytoplasmic transport': ['CellFunction'],\n",
       "  'biomarkers': ['ClinicalAttribute']},\n",
       " 735: {'cerebrospinal fluid': ['BodyPartOrganOrOrganComponent'],\n",
       "  'CSF': ['BodyPartOrganOrOrganComponent'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'in vivo': ['ResearchActivity'],\n",
       "  'dysfunctional': ['PathologicFunction']},\n",
       " 736: {'early detection': ['HealthCareActivity'],\n",
       "  'disease progression': ['PathologicFunction'],\n",
       "  'dynamics': ['NaturalPhenomenonOrProcess'],\n",
       "  'motor activity': ['CellFunction'],\n",
       "  'study': ['ResearchActivity'],\n",
       "  'ALS': ['DiseaseOrSyndrome']},\n",
       " 743: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'impairment': ['Finding'],\n",
       "  'progression': ['PathologicFunction', 'TemporalConcept'],\n",
       "  'neurofilament light chain': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance']},\n",
       " 774: {'Drosophila': ['Eukaryote'],\n",
       "  \"Alzheimer's Disease\": ['DiseaseOrSyndrome'],\n",
       "  'Amyotrophic Lateral Sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'ALS': ['DiseaseOrSyndrome'],\n",
       "  'Ataxia Telangiectasia': ['DiseaseOrSyndrome']},\n",
       " 806: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'Cognitive impairment': ['DiseaseOrSyndrome'],\n",
       "  'MoCA': ['HealthCareActivity'],\n",
       "  'cognitive decline': ['DiseaseOrSyndrome']},\n",
       " 821: {'emergency medical services': ['HealthCareActivity'],\n",
       "  'management': ['HealthCareActivity'],\n",
       "  'diagnosis': ['HealthCareActivity']},\n",
       " 823: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'daily': ['TemporalConcept'],\n",
       "  'ROS': ['BiologicallyActiveSubstance'],\n",
       "  'GSH': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance',\n",
       "   'PharmacologicSubstance']},\n",
       " 827: {'HD': ['DiseaseOrSyndrome']},\n",
       " 875: {'HD': ['DiseaseOrSyndrome'],\n",
       "  'huntingtin gene': ['GeneOrGenome'],\n",
       "  'individuals': ['PopulationGroup'],\n",
       "  'lab': ['HealthCareRelatedOrganization']},\n",
       " 920: {'human': ['Eukaryote'],\n",
       "  \"Alzheimer's disease\": ['DiseaseOrSyndrome'],\n",
       "  'amyotrophic lateral sclerosis': ['DiseaseOrSyndrome'],\n",
       "  'treatment': ['HealthCareActivity'],\n",
       "  'therapy': ['HealthCareActivity']},\n",
       " 942: {'HD': ['DiseaseOrSyndrome']},\n",
       " 960: {'Post-translational modification': ['CellFunction'],\n",
       "  'acetylation': ['CellFunction'],\n",
       "  'glycosylation': ['CellFunction'],\n",
       "  'phosphorylation': ['CellFunction'],\n",
       "  'ubiquitination': ['CellFunction'],\n",
       "  'palmitoylation': ['CellFunction'],\n",
       "  'protein homeostasis': ['CellFunction'],\n",
       "  'protein structure': ['AminoAcidPeptideOrProtein'],\n",
       "  'studies': ['ResearchActivity'],\n",
       "  'AMPK': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'Wnt signaling pathway': ['CellFunction'],\n",
       "  'inhibitor': ['BiologicallyActiveSubstance',\n",
       "   'Chemical',\n",
       "   'PharmacologicSubstance'],\n",
       "  'activator': ['BiologicallyActiveSubstance'],\n",
       "  'protein aggregation': ['CellFunction'],\n",
       "  'alteration': ['Finding'],\n",
       "  'HD': ['DiseaseOrSyndrome'],\n",
       "  'Superoxide dismutase 1': ['AminoAcidPeptideOrProtein'],\n",
       "  'Heat shock proteins': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'PSEN2': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'TARDBP': ['AminoAcidPeptideOrProtein'],\n",
       "  'TAR DNA Binding Protein': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'CK1': ['GeneOrGenome'],\n",
       "  'Casein kinase 1': ['GeneOrGenome', 'AminoAcidPeptideOrProtein'],\n",
       "  'GSK-3 beta': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'Protein kinase A': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'CDK5': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'Cyclin-dependent kinase 5': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'DYRK1A': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'PARKIN': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'FUS': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'Nuclear localization sequence': ['MolecularSequence'],\n",
       "  'NES': ['MolecularSequence'],\n",
       "  'Unfolded protein response': ['CellFunction'],\n",
       "  'Inositol-requiring enzyme 1 alpha': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'PERK': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'Protein kinase R': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'Activating transcription factor 6 alpha': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'Death receptor 5': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'Eukaryotic initiation factor 2 alpha': ['GeneOrGenome'],\n",
       "  'XBP1': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'X-box binding protein 1': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'TRAF2': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'TNF receptor-associated factor 2': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'SP1': ['AminoAcidPeptideOrProtein'],\n",
       "  'Specificity protein 1': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'SP2': ['GeneOrGenome', 'AminoAcidPeptideOrProtein'],\n",
       "  'Specificity protein 2': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'poly ADP ribose polymerase 16': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'CHOP': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'APP': ['GeneOrGenome'],\n",
       "  'Amyloid precursor protein': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'PSEN1': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'Presenilin-1': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'BiP': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'Binding immunoglobulin protein': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'E3 ubiquitin-protein ligase parkin': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'PINK1': ['GeneOrGenome',\n",
       "   'AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'PDI': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'Protein disulfide isomerase': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance'],\n",
       "  'DJ1': ['AminoAcidPeptideOrProtein', 'BiologicallyActiveSubstance'],\n",
       "  'Cyclic AMP-dependent transcription factor': ['AminoAcidPeptideOrProtein',\n",
       "   'BiologicallyActiveSubstance']}}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/usrvol/data/categorical/graphs/final_graphs1.pickle', 'rb') as f:\n",
    "    final_graphs = pkl.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['expansion', 'huntingtin gene', 'Phosphorylation', 'translation initiation factor', 'alteration', 'protein synthesis', 'age', 'factor', 'adult', 'factors', 'maternal', 'diagnosis', 'care', 'primary', 'review', 'genetics', 'pathology'])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_graphs.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comprobamos el tamaño de los grafos obtenidos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle as pkl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "282\n"
     ]
    }
   ],
   "source": [
    "max_len = 0\n",
    "for iterator in range(1, 1001):\n",
    "    try:\n",
    "        with open(f'/usrvol/data/categorical/graphs/final_graphs{iterator}.pickle', 'rb') as f:\n",
    "            final_graphs = pkl.load(f)\n",
    "    except:\n",
    "        continue\n",
    "    for term in final_graphs.keys():\n",
    "        for graph in final_graphs[term]:\n",
    "            if len(graph.nodes) > max_len:\n",
    "                max_len = len(graph.nodes)\n",
    "\n",
    "print(max_len)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Vemos los datos de entrada al modelo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle as pkl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/usrvol/data/categorical/graph_statements_tensors/graph_statementsdev.pickle', 'rb') as f:\n",
    "    graphs = pkl.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "torch.Size([8160, 1, 300])"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graphs[0].size()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Modidicamos el generador de grafos hasta obtener tamaños adecuados"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.10/dist-packages/torch/cuda/__init__.py:546: UserWarning: Can't initialize NVML\n",
      "  warnings.warn(\"Can't initialize NVML\")\n",
      "/usr/local/lib/python3.10/dist-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import pickle as pkl\n",
    "from spacy_utils import *\n",
    "from entities import *\n",
    "from utils import *\n",
    "from graph import *\n",
    "from embeddings import *\n",
    "import networkx as nx\n",
    "from tqdm import tqdm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def expanded_graph(target, entity_tree, TUI_tree, ner, le):\n",
    "    if type(target) is not str and type(target) is not int:\n",
    "        for parent_entity in range(len(target)):\n",
    "            entity_tree.append([])\n",
    "            TUI_tree.append([])\n",
    "            expanded_graph(target[parent_entity], \n",
    "                           entity_tree[parent_entity], \n",
    "                           TUI_tree[parent_entity],\n",
    "                           ner, le)\n",
    "    else:\n",
    "        doc = ner.detected_entities(str(target))\n",
    "        list_linked_entities = ner.linked_entities(doc)\n",
    "        entities, TUIs = le.getNextEntitiesAndTUIs(target, list_linked_entities)\n",
    "        for entity in entities:\n",
    "            entity_tree.append(entity)\n",
    "        for TUI in TUIs:\n",
    "            TUI_tree.append(TUI)\n",
    "       \n",
    "    return entity_tree, TUI_tree\n",
    "\n",
    "def subgraph_generator(entities, graph, obtained_entities, TUIs, limit= 200, total_limit = 1000):\n",
    "    obtained_entities = list(obtained_entities)\n",
    "    entity_list = []\n",
    "    triple_list = []\n",
    "    for entity in entities:\n",
    "        if entity in graph.nodes():\n",
    "            counter_in = 0\n",
    "            for triple in graph.in_edges(entity, data=True):\n",
    "                if triple[0] not in obtained_entities:\n",
    "                    if counter_in < limit:\n",
    "                        entity_list.append(triple[0])\n",
    "                        obtained_entities.append(triple[0])\n",
    "                        triple_list.append((triple[0], triple[2]['rel'], triple[1]))\n",
    "                        counter_in +=1\n",
    "                        break\n",
    "            counter_out = 0\n",
    "            for triple in graph.out_edges(entity, data=True):\n",
    "                if triple[1] not in obtained_entities:\n",
    "                    if counter_out < limit:\n",
    "                        entity_list.append(triple[1])\n",
    "                        obtained_entities.append(triple[1])\n",
    "                        triple_list.append((triple[0], triple[2]['rel'], triple[1]))\n",
    "                        counter_out +=1\n",
    "                        break\n",
    "    obtained_entities = list(set(obtained_entities))\n",
    "    entity_list = list(set(entity_list))\n",
    "    if len(obtained_entities) > total_limit:\n",
    "        obtained_entities = obtained_entities[:total_limit]\n",
    "        entity_list = entity_list[:total_limit]\n",
    "        for triple in triple_list:\n",
    "            if triple[0] not in obtained_entities:\n",
    "                triple_list.remove(triple)\n",
    "                continue\n",
    "            if triple[1] not in obtained_entities:\n",
    "                triple_list.remove(triple)\n",
    "    return entity_list, triple_list, obtained_entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "with open('/usrvol/mappings/CUIsToInt.pickle', 'rb') as f:\n",
    "    CuisToInt = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/graph.pickle', 'rb') as f:\n",
    "    G = pkl.load(f)    \n",
    "\n",
    "with open('/usrvol/mappings/NamestoCUIFinal.pickle', 'rb') as f:\n",
    "    NamesToCUIs = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/IntToCUIs.pickle', 'rb') as f:\n",
    "    IntToCUIs = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/CUItoPreferredNamesFinal.pickle', 'rb') as f:\n",
    "    CUIToNames = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/IntToNames.pickle', 'rb') as f:\n",
    "    IntToNames = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/cuiToTUIs.pickle', 'rb') as f:\n",
    "    CUItoTUIs = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/TUIsToInt.pickle', 'rb') as f:\n",
    "    TUIToInt = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/RelationtoIdentifierTotal.pickle', 'rb') as f:\n",
    "    RelationToIdentifier = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/mapping_definitivo.pickle', 'rb') as f:\n",
    "    labelMapping = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/TUItoSemantic.pickle', 'rb') as f:\n",
    "    TUItoText = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/classes.pickle', 'rb') as f:\n",
    "    classes = pkl.load(f)\n",
    "\n",
    "with open('/usrvol/mappings/classesToTextDict.pickle', 'rb') as f:\n",
    "    classesToText = pkl.load(f)\n",
    "\n",
    "mappers = (CuisToInt, IntToCUIs, NamesToCUIs, CUIToNames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.10/dist-packages/spacy/language.py:2141: FutureWarning: Possible set union at position 6328\n",
      "  deserializers[\"tokenizer\"] = lambda p: self.tokenizer.from_disk(  # type: ignore[union-attr]\n",
      "/usr/local/lib/python3.10/dist-packages/sklearn/base.py:347: InconsistentVersionWarning: Trying to unpickle estimator TfidfTransformer from version 1.1.2 when using version 1.3.0. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.10/dist-packages/sklearn/base.py:347: InconsistentVersionWarning: Trying to unpickle estimator TfidfVectorizer from version 1.1.2 when using version 1.3.0. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# función recursiva que nos expande el grafo a partir de un target devolviendo un diccionario\n",
    "# con sus conexiones\n",
    "ner = Ner(spacy_model='en_core_sci_lg', \n",
    "          pipeline_linker='scispacy_linker', \n",
    "          resolve_abbreviations=True, \n",
    "          ontology=\"umls\", \n",
    "          threshold=\"0.85\", \n",
    "          max_entities_per_mention='100')\n",
    "\n",
    "le = Linked_Entities(resolve_abbreviations=True, ontology=\"umls\", threshold=0.85, max_entities_per_mention=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "text_path = '/usrvol/ISABIAL/Anotaciones_v2_ISABIAL/txt'\n",
    "entries = os.listdir(text_path)\n",
    "annotations_dict = {}\n",
    "vocab_mapper = {}\n",
    "iterator = 0\n",
    "counter_mapper = 0\n",
    "desambiguation_entities_number = 6000000\n",
    "classes_number = 7000000\n",
    "NamesToInt = {}\n",
    "maximum_hopes = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Este solo lo ejecutamos una vez\n",
    "for class_ in classes:\n",
    "    IntToNames[classes_number] = classesToText[class_]\n",
    "    NamesToInt[class_] = classes_number\n",
    "    vocab_mapper[classes_number] = counter_mapper\n",
    "    classes_number += 1\n",
    "    counter_mapper += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'tqdm' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/usrvol/GitHub2/debug_categorical.ipynb Cell 47\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f72656c617865645f63686174656c6574227d/usrvol/GitHub2/debug_categorical.ipynb#X65sdnNjb2RlLXJlbW90ZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mfor\u001b[39;00m file_path \u001b[39min\u001b[39;00m tqdm(\u001b[39mrange\u001b[39m(\u001b[39m1000\u001b[39m)):\n\u001b[1;32m      <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f72656c617865645f63686174656c6574227d/usrvol/GitHub2/debug_categorical.ipynb#X65sdnNjb2RlLXJlbW90ZQ%3D%3D?line=1'>2</a>\u001b[0m     iterator \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m \u001b[39m1\u001b[39m\n\u001b[1;32m      <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f72656c617865645f63686174656c6574227d/usrvol/GitHub2/debug_categorical.ipynb#X65sdnNjb2RlLXJlbW90ZQ%3D%3D?line=2'>3</a>\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mfile number: \u001b[39m\u001b[39m{\u001b[39;00mfile_path\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m )\n",
      "\u001b[0;31mNameError\u001b[0m: name 'tqdm' is not defined"
     ]
    }
   ],
   "source": [
    "for file_path in tqdm(range(1000)):\n",
    "    iterator += 1\n",
    "    print(f\"file number: {file_path}\" )\n",
    "    if f\"text{file_path}.txt\" in entries:\n",
    "        file = f\"text{file_path}.txt\"\n",
    "        with open(f\"/usrvol/ISABIAL/Anotaciones_v2_ISABIAL/txt/{file}\", 'r') as f:\n",
    "            lines = f.readlines()\n",
    "        text = lines[0]\n",
    "        candidates_dict = {}\n",
    "        doc = ner.detected_entities(text)\n",
    "        list_linked_entities = ner.linked_entities(doc)\n",
    "        candidates, candidates_TUIs = le.candidatesChecker(list_linked_entities)\n",
    "        \n",
    "        # Con esto evitamos repetir candidatos si aparecen varias veces en el texto\n",
    "        new_candidates = []\n",
    "        for candidate in candidates:\n",
    "            if candidate not in new_candidates:\n",
    "                new_candidates.append(candidate)\n",
    "        candidates = new_candidates\n",
    "        \n",
    "        \n",
    "        entity_list, TUI_list = le.getEntitiesAndTUIs(list_linked_entities)\n",
    "        final_graphs = {}\n",
    "        labels = {}\n",
    "\n",
    "        # Creamos una lista de Nombres de entidades a desambiguar mapeandolas a un entero\n",
    "        for candidate in range(len(candidates)):     \n",
    "            if not candidates[candidate] in NamesToInt.keys():\n",
    "                IntToNames[desambiguation_entities_number] = candidates[candidate]\n",
    "                NamesToInt[candidates[candidate]] = desambiguation_entities_number\n",
    "                vocab_mapper[desambiguation_entities_number] = counter_mapper\n",
    "                desambiguation_entities_number += 1\n",
    "                counter_mapper +=1\n",
    "            \n",
    "            else:\n",
    "                # Si se cumple esta condición quiere decir que el término aparece tanto como candidato como en el vocabulario original\n",
    "                if NamesToInt[candidates[candidate]] not in vocab_mapper.keys():\n",
    "                    vocab_mapper[NamesToInt[candidates[candidate]]] = counter_mapper\n",
    "                    counter_mapper +=1\n",
    "\n",
    "        # iteramos sobre cada candidato teniendo ya el vocabulario de los candidatos   \n",
    "        for candidate in range(len(candidates)):\n",
    "            graph_entities = {'first_entities': [], 'last_entities': [], 'TUIs': []}\n",
    "            obtained_entities = []\n",
    "            TUIs = candidates_TUIs[candidate]\n",
    "            forward_triples_list = []\n",
    "            backward_triples_list = []\n",
    "            connected_entities, _ = expanded_graph(candidates[candidate], [], [], ner, le)\n",
    "\n",
    "            # Nos quedamos solo con las entidades que estén en el vocabulario según la base de datos\n",
    "            newconnected_entities = []\n",
    "            for entity in range(len(connected_entities)):\n",
    "                    if connected_entities[entity] in NamesToCUIs.keys():\n",
    "                        newconnected_entities.append(CuisToInt[NamesToCUIs[connected_entities[entity].lower()]])\n",
    "                        \n",
    "                        # Añadimos estas entidades al vocabulario\n",
    "                        if not connected_entities[entity] in NamesToInt.keys():\n",
    "                            IntToNames[desambiguation_entities_number] = connected_entities[entity]\n",
    "                            NamesToInt[connected_entities[entity]] = desambiguation_entities_number\n",
    "                            vocab_mapper[desambiguation_entities_number] = counter_mapper\n",
    "                            desambiguation_entities_number += 1\n",
    "                            counter_mapper +=1\n",
    "            \n",
    "            connected_entities = newconnected_entities\n",
    "            graph_entities['first_entities'] = connected_entities\n",
    "  \n",
    "            # formamos el grafo desde las entidades iniciales\n",
    "            for hop in range(maximum_hopes):\n",
    "                connected_entities, triples, obtained_entities = subgraph_generator(connected_entities, G, obtained_entities, TUIs)\n",
    "                forward_triples_list.extend(triples)\n",
    "                forward_triples_list = list(set(forward_triples_list))\n",
    "            Rconnected_entities, RTUIs = le.otherRelatedEntities(candidates[candidate], \n",
    "                                                        candidates, \n",
    "                                                        candidates_TUIs, \n",
    "                                                        entity_list, TUI_list)\n",
    "            \n",
    "            # Nos quedamos solo con las entidades que estén en el vocabulario según la base de datos\n",
    "            newRconnected_entities = []\n",
    "            for entity in range(len(Rconnected_entities)):\n",
    "                if Rconnected_entities[entity] in NamesToCUIs.keys():\n",
    "                    if not Rconnected_entities[entity] in NamesToInt.keys():\n",
    "                        IntToNames[desambiguation_entities_number] = Rconnected_entities[entity]\n",
    "                        NamesToInt[Rconnected_entities[entity]] = desambiguation_entities_number\n",
    "                        vocab_mapper[desambiguation_entities_number] = counter_mapper\n",
    "                        desambiguation_entities_number += 1\n",
    "                        counter_mapper +=1\n",
    "\n",
    "                    if NamesToCUIs[Rconnected_entities[entity].lower()] not in CuisToInt.keys():\n",
    "                        CuisToInt[NamesToCUIs[Rconnected_entities[entity].lower()]] = counter_mapper\n",
    "                    newRconnected_entities.append(CuisToInt[NamesToCUIs[Rconnected_entities[entity].lower()]])\n",
    "            Rconnected_entities = newRconnected_entities\n",
    "            graph_entities['last_entities'] = Rconnected_entities\n",
    "            graph_entities['TUIs'] = RTUIs\n",
    "            statements_graphs = []\n",
    "           \n",
    "            # añadimos los TUIs a labels para usarlos como etiquetas\n",
    "            for TUI in TUIs:\n",
    "                if not candidates[candidate] in labels.keys():\n",
    "                    labels[candidates[candidate]] = [labelMapping[TUI]]\n",
    "                else:\n",
    "                    labels[candidates[candidate]].append(labelMapping[TUI])\n",
    "\n",
    "            # Formamos todas las variantes de grafo desde las entidades finales\n",
    "            for class_ in classes:\n",
    "                if class_ in labels.keys():\n",
    "                    graph_last_entities = []\n",
    "\n",
    "                    # Si la entidad final de la lista de entidades contiene entre sus TUIs el TUI objetivo...\n",
    "                    for entity in range(len(graph_entities['last_entities'])):\n",
    "                        if TUI in graph_entities['TUIs'][entity]:\n",
    "                            # Se añade a graph_last_entities para formar este grafo en particular\n",
    "                            graph_last_entities.append(graph_entities['last_entities'][entity])\n",
    "                    if len(graph_last_entities) < 30:\n",
    "                        final_entities = graph_last_entities\n",
    "                    else:\n",
    "                        final_entities = graph_last_entities[:30]\n",
    "\n",
    "                    print(\"final_entities added: \", len(final_entities))\n",
    "                    # se obtienen las tripletas inversas para este TUI en particular\n",
    "                    for hop in range(maximum_hopes):      \n",
    "                        graph_last_entities, triples, obtained_entities = subgraph_generator(graph_last_entities, G, obtained_entities, TUIs)\n",
    "                        backward_triples_list.extend(triples)\n",
    "                        backward_triples_list = list(set(backward_triples_list))\n",
    "                    triples_list = forward_triples_list + backward_triples_list\n",
    "                    graph = nx.MultiDiGraph()\n",
    "                    target = candidates[candidate]\n",
    "                    initial_entities = graph_entities['first_entities']\n",
    "                    attrs = set()\n",
    "                    for triple in triples_list:\n",
    "                        head = triple[0]\n",
    "                        # Añadimos todas las entidades y relaciones de las tripletas de cada grafo al vocabulario si no existen aún\n",
    "                        if not IntToNames[head] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[head]] = head\n",
    "                            vocab_mapper[head] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "                        elif not head in vocab_mapper.keys():\n",
    "                            head = NamesToInt[IntToNames[head]]\n",
    "                        tail = triple[2]\n",
    "                        # Idem\n",
    "                        if not IntToNames[tail] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[tail]] = tail\n",
    "                            vocab_mapper[tail] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "                        elif not tail in vocab_mapper.keys():\n",
    "                            tail = NamesToInt[IntToNames[tail]]   \n",
    "                        graph.add_node(vocab_mapper[head], TUIs='None', node_type=2, node_score=0)\n",
    "                        graph.add_node(vocab_mapper[tail], TUIs='None', node_type=2, node_score=0)\n",
    "                        rel = triple[1]\n",
    "                        weight = 1.\n",
    "                        graph.add_edge(vocab_mapper[head], vocab_mapper[tail], rel=rel, weight=weight)\n",
    "                        attrs.add((vocab_mapper[head], vocab_mapper[tail], rel))\n",
    "                    # añadimos el nodo inicial y sus conexiones\n",
    "                    graph.add_node(vocab_mapper[NamesToInt[target]], TUIs=TUIs, node_type=0, node_score=0)\n",
    "                    for entity in initial_entities:\n",
    "                        if not IntToNames[entity] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[entity]] = entity\n",
    "                            vocab_mapper[entity] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "                        elif not entity in vocab_mapper.keys():\n",
    "                            entity = NamesToInt[IntToNames[entity]]\n",
    "                        if not entity in graph.nodes:\n",
    "                            try:\n",
    "                                T = CUItoTUIs[IntToCUIs[entity]]\n",
    "                            except:\n",
    "                                T = 'NOT FOUND'\n",
    "                        graph.add_node(vocab_mapper[entity], TUIs=[T], node_type=2, node_score=0)\n",
    "                        graph.add_edge(vocab_mapper[NamesToInt[target]], vocab_mapper[entity], rel=RelationToIdentifier['meaning_of_concept'], weight=1.)\n",
    "                        \n",
    "                    \n",
    "                    # añadimos el nodo final y sus conexiones\n",
    "                    graph.add_node(vocab_mapper[NamesToInt[class_]], TUIs=['None'], node_type=1, node_score=0)\n",
    "                    for entity in final_entities:\n",
    "                        if not IntToNames[entity] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[entity]] = entity\n",
    "                            vocab_mapper[entity] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "                        elif not entity in vocab_mapper.keys():\n",
    "                            entity = NamesToInt[IntToNames[entity]]\n",
    "                        if not entity in graph.nodes:\n",
    "                            try:\n",
    "                                T = CUItoTUIs[IntToCUIs[entity]]\n",
    "                            except:\n",
    "                                T = 'NOT FOUND'\n",
    "                        graph.add_node(vocab_mapper[entity], TUIs=['None'], node_type=2, node_score=0)\n",
    "                        graph.add_edge(vocab_mapper[entity], vocab_mapper[NamesToInt[class_]], rel=RelationToIdentifier['belongs_to'], weight=1.)\n",
    "                    statements_graphs.append(graph)\n",
    "                    print(graph.number_of_nodes())\n",
    "                    backward_triples_list = []\n",
    "                else:\n",
    "                    triples_list = forward_triples_list\n",
    "                    graph = nx.MultiDiGraph()\n",
    "                    target = candidates[candidate]\n",
    "                    initial_entities = graph_entities['first_entities']\n",
    "                    print(\"initial_entities added: \", len(initial_entities))\n",
    "                    attrs = set()\n",
    "                    for triple in triples_list:             \n",
    "                        head = triple[0]\n",
    "                        # Añadimos todas las entidades y relaciones de las tripletas de cada grafo al vocabulario si no existen aún\n",
    "                        if not IntToNames[head] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[head]] = head\n",
    "                            vocab_mapper[head] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "                        # Esto quiere decir que el término aparece tanto como candidato como en el vocabulario original, teniendo dos números diferentes que vamos a mapear a uno solo\n",
    "                        elif not head in vocab_mapper.keys():\n",
    "                            head = NamesToInt[IntToNames[head]]       \n",
    "                        tail = triple[2]\n",
    "                        # Idem\n",
    "                        if not IntToNames[tail] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[tail]] = tail\n",
    "                            vocab_mapper[tail] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "                        # Esto quiere decir que el término aparece tanto como candidato como en el vocabulario original, teniendo dos números diferentes que vamos a mapear a uno solo\n",
    "                        elif not tail in vocab_mapper.keys():\n",
    "                            tail = NamesToInt[IntToNames[tail]]\n",
    "\n",
    "                        graph.add_node(vocab_mapper[head], TUIs='none', node_type=2, node_score=0)\n",
    "                        graph.add_node(vocab_mapper[tail], TUIs='none', node_type=2, node_score=0)\n",
    "                        rel = triple[1]\n",
    "                        weight = 1.\n",
    "                        graph.add_edge(vocab_mapper[head], vocab_mapper[tail], rel=rel, weight=weight)\n",
    "                        attrs.add((vocab_mapper[head], vocab_mapper[tail], rel))\n",
    "                    # añadimos el nodo inicial y sus conexiones\n",
    "                    graph.add_node(vocab_mapper[NamesToInt[target]], TUIs=TUIs, node_type=0, node_score=0)\n",
    "                    for entity in initial_entities:\n",
    "                        if not IntToNames[entity] in NamesToInt.keys():\n",
    "                            NamesToInt[IntToNames[entity]] = entity\n",
    "                            vocab_mapper[entity] = counter_mapper\n",
    "                            counter_mapper +=1\n",
    "                        # Esto quiere decir que el término aparece tanto como candidato como en el vocabulario original, teniendo dos números diferentes que vamos a mapear a uno solo\n",
    "                        elif not entity in vocab_mapper.keys():\n",
    "                            entity = NamesToInt[IntToNames[entity]]\n",
    "\n",
    "                        if not entity in graph.nodes:\n",
    "                            try:\n",
    "                                T = CUItoTUIs[IntToCUIs[entity]]\n",
    "                            except:\n",
    "                                T = 'NOT FOUND'     \n",
    "                        graph.add_node(vocab_mapper[entity], TUIs=[T], node_type=2, node_score=0)\n",
    "                        graph.add_edge(vocab_mapper[NamesToInt[target]], vocab_mapper[entity], rel=RelationToIdentifier['meaning_of_concept'], weight=1.)\n",
    "                    # añadimos el nodo final y sin conexiones porque no es su clase\n",
    "                    graph.add_node(vocab_mapper[NamesToInt[class_]], TUIs=['None'], node_type=1, node_score=0)\n",
    "                    statements_graphs.append(graph)\n",
    "                    print(graph.number_of_nodes())           \n",
    "            final_graphs[candidates[candidate]] = statements_graphs"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
